// @webex/plugin-meetings/dist/webex.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

require('@webex/plugin-authorization');

require('@webex/internal-plugin-wdm');

require('@webex/plugin-logger');

require('@webex/plugin-people');

require('@webex/plugin-rooms');

var _webexCore = require('@webex/webex-core');

var _webexCore2 = _interopRequireDefault(_webexCore);

var _merge = require('lodash/merge');

var _merge2 = _interopRequireDefault(_merge);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Webex = _webexCore2.default.extend({
  webex: true,
  version: '1.72.6'
});

Webex.version = '1.72.6';

Webex.init = function init() {
  var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var config = (0, _merge2.default)({}, {
    hydra: process.env.HYDRA_SERVICE_URL || 'https://api.ciscospark.com/v1',
    credentials: {
      clientType: 'confidential'
    },
    device: {
      ephemeral: true
    },
    storage: {
      boundedAdapter: _webexCore.MemoryStoreAdapter,
      unboundedAdapter: _webexCore.MemoryStoreAdapter
    }
  }, attrs.config);

  return new Webex((0, _merge2.default)({}, { config: config }, attrs));
};

exports.default = Webex;
