// @webex/webex-core/dist/interceptors/auth.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _httpCore = require('@webex/http-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
var AuthInterceptor = function (_Interceptor) {
  (0, _inherits3.default)(AuthInterceptor, _Interceptor);

  function AuthInterceptor() {
    (0, _classCallCheck3.default)(this, AuthInterceptor);
    return (0, _possibleConstructorReturn3.default)(this, (AuthInterceptor.__proto__ || (0, _getPrototypeOf2.default)(AuthInterceptor)).apply(this, arguments));
  }

  (0, _createClass3.default)(AuthInterceptor, [{
    key: 'onRequest',


    /**
     * @see {@link Interceptor#onRequest}
     * @param {Object} options
     * @returns {Object}
     */
    value: function onRequest(options) {
      var _this2 = this;

      options.headers = options.headers || {};

      // If Authorizations is already set, don't overwrite it
      if ('authorization' in options.headers || 'auth' in options) {
        // If Authorization is set to null, false, or undefined, delete it to
        // prevent a CORS preflight.
        if (!options.headers.authorization) {
          (0, _deleteProperty2.default)(options.headers, 'authorization');
        }

        return _promise2.default.resolve(options);
      }

      return this.requiresCredentials(options).then(function (requires) {
        if (!requires) {
          return options;
        }

        return _this2.webex.credentials.getUserToken().then(function (token) {
          options.headers.authorization = token.toString();

          return options;
        });
      });
    }

    /**
     * Determines if the provided options object needs an auth header
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'requiresCredentials',
    value: function requiresCredentials(options) {
      var _this3 = this;

      if (options.uri && (options.uri.includes(this.webex.config.device.preDiscoveryServices.hydra) || options.uri.includes(this.webex.config.device.preDiscoveryServices.hydraServiceUrl))) {
        return _promise2.default.resolve(true);
      }

      if (!this.webex.internal.device) {
        return _promise2.default.resolve(false);
      }

      return this.webex.internal.device.isSpecificService('hydra', options.uri).then(function (isHydra) {
        if (isHydra) {
          return true;
        }

        if (options.service) {
          return _this3.webex.internal.device.isService(options.service);
        }
        if (options.uri) {
          // If service Url then return or else check if whitelisted
          return _this3.webex.internal.device.isServiceUrl(options.uri).then(function (isServiceUrl) {
            if (isServiceUrl) {
              return true;
            }

            if (options.addAuthHeader === undefined && _this3.webex.config.device.validateDomains) {
              options.addAuthHeader = true;
            }

            // returns true if uri is in whitelistedServiceDomains and is requested
            if (options.addAuthHeader) {
              var matchingDomains = _this3.webex.config.device.whitelistedServiceDomains.filter(function (domain) {
                return options.uri.includes(domain);
              });

              if (matchingDomains.length) {
                return _promise2.default.resolve(true);
              }
            }

            return false;
          });
        }

        return false;
      });
    }

    /**
     * @see {@link Interceptor#onResponseError}
     * @param {Object} options
     * @param {Error} reason
     * @returns {Object}
     */

  }, {
    key: 'onResponseError',
    value: function onResponseError(options, reason) {
      var _this4 = this;

      return this.shouldAttemptReauth(reason, options).then(function (shouldAttemptReauth) {
        if (shouldAttemptReauth) {
          _this4.webex.logger.info('auth: received 401, attempting to reauthenticate');

          if (reason.options.headers) {
            (0, _deleteProperty2.default)(reason.options.headers, 'authorization');
          }

          if (_this4.webex.credentials.canRefresh) {
            return _this4.webex.credentials.refresh().then(function () {
              return _this4.replay(options);
            });
          }
        }

        return _promise2.default.reject(reason);
      });
    }

    /**
     * Replays the request
     * @param {Object} options
     * @returns {Object}
     */

  }, {
    key: 'replay',
    value: function replay(options) {
      if (options.replayCount) {
        options.replayCount += 1;
      } else {
        options.replayCount = 1;
      }

      if (options.replayCount > this.webex.config.maxAuthenticationReplays) {
        this.webex.logger.error('auth: failed after ' + this.webex.config.maxAuthenticationReplays + ' replay attempts');

        return _promise2.default.reject(new Error('Failed after ' + this.webex.config.maxAuthenticationReplays + ' replay attempts'));
      }

      this.webex.logger.info('auth: replaying request ' + options.replayCount + ' time');

      return this.webex.request(options);
    }

    /**
     * Indicates whether or not the current request should refresh its access
     * token in event of a 401
     * @param {Error} reason
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'shouldAttemptReauth',
    value: function shouldAttemptReauth(reason, options) {
      if (options && options.shouldRefreshAccessToken === false) {
        return _promise2.default.resolve(false);
      }

      if (reason.statusCode === 401) {
        return _promise2.default.resolve(true);
      }

      return _promise2.default.resolve(false);
    }
  }], [{
    key: 'create',

    /**
     * @returns {AuthInterceptor}
     */
    value: function create() {
      return new AuthInterceptor({ webex: this });
    }
  }]);
  return AuthInterceptor;
}(_httpCore.Interceptor); /*!
                           * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                           */

exports.default = AuthInterceptor;
