// @webex/webex-core/dist/interceptors/rate-limit.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _map = require('babel-runtime/core-js/map');

var _map2 = _interopRequireDefault(_map);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _weakMap = require('babel-runtime/core-js/weak-map');

var _weakMap2 = _interopRequireDefault(_weakMap);

var _httpCore = require('@webex/http-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// contains the system time in milliseconds at which the retry after associated with a 429 expires
// mapped by the API name, e.g.: idbroker.webex.com/doStuff would be mapped as 'doStuff'
var rateLimitExpiryTime = new _weakMap2.default();

// extracts the common identity API being called
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

var idBrokerRegex = /.*(idbroker|identity)(bts)?.ciscospark.com\/([^/]+)/;

/**
 * @class
 */

var RateLimitInterceptor = function (_Interceptor) {
  (0, _inherits3.default)(RateLimitInterceptor, _Interceptor);
  (0, _createClass3.default)(RateLimitInterceptor, null, [{
    key: 'create',

    /**
     * @returns {RateLimitInterceptor}
     */
    value: function create() {
      return new RateLimitInterceptor({ webex: this });
    }

    /**
    * constructor
    * @param {mixed} args
    * @returns {Exception}
    */

  }]);

  function RateLimitInterceptor() {
    var _ref;

    (0, _classCallCheck3.default)(this, RateLimitInterceptor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = RateLimitInterceptor.__proto__ || (0, _getPrototypeOf2.default)(RateLimitInterceptor)).call.apply(_ref, [this].concat(args)));

    rateLimitExpiryTime.set(_this, new _map2.default());
    return _this;
  }

  /**
   * @see {@link Interceptor#onRequest}
   * @param {Object} options
   * @returns {Object}
   */


  (0, _createClass3.default)(RateLimitInterceptor, [{
    key: 'onRequest',
    value: function onRequest(options) {
      if (this.isRateLimited(options.uri)) {
        return _promise2.default.reject(new Error('API rate limited ' + options.uri));
      }

      return _promise2.default.resolve(options);
    }

    /**
     * @see {@link Interceptor#onResponseError}
     * @param {Object} options
     * @param {Error} reason
     * @returns {Object}
     */

  }, {
    key: 'onResponseError',
    value: function onResponseError(options, reason) {
      if (reason.statusCode === 429 && (options.uri.includes('idbroker') || options.uri.includes('identity'))) {
        // set the retry after in the map, setting to milliseconds
        this.setRateLimitExpiry(options.uri, this.extractRetryAfterTime(options));
      }

      return _promise2.default.reject(reason);
    }

    /**
     * @param {object} options associated with the request
     * @returns {number} retry after time in milliseconds
     */

  }, {
    key: 'extractRetryAfterTime',
    value: function extractRetryAfterTime(options) {
      // 1S * 1K === 1MS
      var milliMultiplier = 1000;
      var retryAfter = options.headers['retry-after'] || null;

      // set 60 retry if no usable time defined
      if (retryAfter === null || retryAfter <= 0) {
        return 60 * milliMultiplier;
      }
      // set max to 3600 S (1 hour) if greater than 1 hour
      if (retryAfter > 3600) {
        return 3600 * milliMultiplier;
      }

      return retryAfter * milliMultiplier;
    }

    /**
     * Set the system time at which the rate limiting
     * will expire in the rateLimitExpiryTime map.
     * Assumes retryAfter is in milliseconds
     * @param {string} uri API issuing the rate limiting
     * @param {number} retryAfter milliseconds until rate limiting expires
     * @returns {bool} true is value was successfully set
     */

  }, {
    key: 'setRateLimitExpiry',
    value: function setRateLimitExpiry(uri, retryAfter) {
      var apiName = this.getApiName(uri);

      if (!apiName) {
        return false;
      }

      var currTimeMilli = new Date().getTime();
      var expiry = currTimeMilli + retryAfter;
      var dict = rateLimitExpiryTime.get(this);

      return dict.set(apiName, expiry);
    }

    /**
     * returns true if the API is currently rate limited
     * @param {string} uri
     * @returns {Boolean} indicates whether or not the API is rate currently rate limited
     */

  }, {
    key: 'getRateLimitStatus',
    value: function getRateLimitStatus(uri) {
      var apiName = this.getApiName(uri);

      if (!apiName) {
        return false;
      }

      var currTimeMilli = new Date().getTime();
      var dict = rateLimitExpiryTime.get(this);
      var expiryTime = dict.get(apiName);

      // if no rate limit expiry has been registered in the map, return false.
      if (expiryTime === undefined) {
        return false;
      }

      // return true, indicating rate limiting, if the system time is less than the expiry time
      return currTimeMilli < dict.get(apiName);
    }

    /**
     * split the URI and returns the API name of idBroker
     * @param {string} uri
     * @returns {string}
     */

  }, {
    key: 'getApiName',
    value: function getApiName(uri) {
      if (!uri) {
        return null;
      }

      var results = uri.match(idBrokerRegex);

      if (!results) {
        return null;
      }

      // group 0 = full match of URL, group 1 = identity or idbroker base
      // group 2 = api name
      return results[2];
    }

    /**
     * check URI against list of currently rate limited
     * URIs, and determines if retry-after
     * @param {String} uri pattern to check
     * @returns {bool}
     */

  }, {
    key: 'isRateLimited',
    value: function isRateLimited(uri) {
      // determine if the URI is associated with a common identity API
      if (uri && (uri.includes('idbroker') || uri.includes('identity'))) {
        return this.getRateLimitStatus(uri);
      }

      return false;
    }
  }]);
  return RateLimitInterceptor;
}(_httpCore.Interceptor);

exports.default = RateLimitInterceptor;
