// @webex/webex-core/dist/plugins/logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webexCore = require('@webex/webex-core');

var precedence = {
  error: ['log'],
  warn: ['error', 'log'],
  info: ['log'],
  debug: ['info', 'log'],
  trace: ['debug', 'info', 'log']
};

/**
 * Assigns the specified console method to Logger; uses `precedence` to fallback
 * to other console methods if the current environment doesn't provide the
 * specified level.
 * @param {string} level
 * @returns {Function}
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

function wrapConsoleMethod(level) {
  /* eslint no-console: [0] */
  var impls = precedence[level];

  if (impls) {
    impls = impls.slice();
    while (!console[level]) {
      level = impls.pop();
    }
  }

  return function wrappedConsoleMethod() {
    var _console;

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    /* eslint no-invalid-this: [0] */
    /* istanbul ignore if */
    if (process.env.NODE_ENV === 'test' && this.webex && this.webex.internal.device && this.webex.internal.device.url) {
      args.unshift(this.webex.internal.device.url.slice(-3));
    }
    (_console = console)[level].apply(_console, args);
  };
}

var Logger = _webexCore.WebexPlugin.extend({
  namespace: 'Logger',
  error: wrapConsoleMethod('error'),
  warn: wrapConsoleMethod('warn'),
  log: wrapConsoleMethod('log'),
  info: wrapConsoleMethod('info'),
  debug: wrapConsoleMethod('debug'),
  trace: wrapConsoleMethod('trace'),
  version: '1.72.6'
});

(0, _webexCore.registerPlugin)('logger', Logger);

exports.default = Logger;
