// @webex/internal-plugin-encryption/dist/encryption.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _webexCore = require('@webex/webex-core');

var _common = require('@webex/common');

var _events = require('events');

var _nodeJose = require('node-jose');

var _nodeJose2 = _interopRequireDefault(_nodeJose);

var _nodeScr = require('node-scr');

var _nodeScr2 = _interopRequireDefault(_nodeScr);

var _ensureBuffer = require('./ensure-buffer');

var _ensureBuffer2 = _interopRequireDefault(_ensureBuffer);

var _kms = require('./kms');

var _kms2 = _interopRequireDefault(_kms);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Encryption = _webexCore.WebexPlugin.extend({
  children: {
    kms: _kms2.default
  },

  namespace: 'Encryption',

  processKmsMessageEvent: function processKmsMessageEvent(event) {
    return this.kms.processKmsMessageEvent(event);
  },
  decryptBinary: function decryptBinary(scr, buffer) {
    return (0, _ensureBuffer2.default)(buffer).then(function (b) {
      /* istanbul ignore if */
      if (buffer.length === 0 || buffer.byteLength === 0) {
        return _promise2.default.reject(new Error('Attempted to decrypt zero-length buffer'));
      }

      return scr.decrypt(b);
    });
  },


  /**
   * Decrypt a SCR (Secure Content Resource) using the supplied key uri.
   *
   * @param {string} key - The uri of a key stored in KMS
   * @param {Object} cipherScr - An encrypted SCR
   * @param {Object} options
   * @param {string} options.onBehalfOf - Fetch the KMS key on behalf of another user (using the user's UUID), active user requires the 'spark.kms_orgagent' role
   * @returns {Object} Decrypted SCR
   */
  decryptScr: function decryptScr(key, cipherScr, options) {
    return this.getKey(key, options).then(function (k) {
      return _nodeScr2.default.fromJWE(k.jwk, cipherScr);
    });
  },


  /**
   * Decrypt text using the supplied key uri.
   *
   * @param {string} key - The uri of a key stored in KMS
   * @param {string} ciphertext - Encrypted text
   * @param {Object} options
   * @param {string} options.onBehalfOf - Fetch the KMS key on behalf of another user (using the user's UUID), active user requires the 'spark.kms_orgagent' role
   * @returns {string} Decrypted plaintext
   */
  decryptText: function decryptText(key, ciphertext, options) {
    return this.getKey(key, options).then(function (k) {
      return _nodeJose2.default.JWE.createDecrypt(k.jwk).decrypt(ciphertext).then(function (result) {
        return result.plaintext.toString();
      });
    });
  },
  download: function download(scr) {
    var _this = this;

    /* istanbul ignore if */
    if (!scr.loc) {
      return _promise2.default.reject(new Error('`scr.loc` is required'));
    }

    var shunt = new _events.EventEmitter();
    var promise = this._fetchDownloadUrl(scr).then(function (uri) {
      var options = {
        method: 'GET',
        uri: uri,
        responseType: 'buffer'
      };

      var ret = _this.request(options);

      (0, _common.transferEvents)('progress', options.download, shunt);

      return ret;
    }).then(function (res) {
      return _this.decryptBinary(scr, res.body);
    });

    (0, _common.proxyEvents)(shunt, promise);

    return promise;
  },
  _fetchDownloadUrl: function _fetchDownloadUrl(scr) {
    var _this2 = this;

    this.logger.info('encryption: retrieving download url for encrypted file');

    if (process.env.NODE_ENV !== 'production' && scr.loc.includes('localhost')) {
      this.logger.info('encryption: bypassing webex files because this looks to be a test file on localhost');

      return _promise2.default.resolve(scr.loc);
    }

    return this.request({
      method: 'POST',
      service: 'files',
      resource: 'download/endpoints',
      body: {
        endpoints: [scr.loc]
      }
    }).then(function (res) {
      var url = res.body.endpoints[scr.loc];

      if (!url) {
        _this2.logger.warn('encryption: could not determine download url for `scr.loc`; attempting to download `scr.loc` directly');

        return scr.loc;
      }
      _this2.logger.info('encryption: retrieved download url for encrypted file');

      return url;
    });
  },
  encryptBinary: function encryptBinary(file) {
    return (0, _ensureBuffer2.default)(file).then(function (buffer) {
      return _nodeScr2.default.create().then(function (scr) {
        return scr.encrypt(buffer).then(_ensureBuffer2.default)
        // eslint-disable-next-line max-nested-callbacks
        .then(function (cdata) {
          return { scr: scr, cdata: cdata };
        });
      });
    });
  },


  /**
   * Encrypt a SCR (Secure Content Resource) using the supplied key uri.
   *
   * @param {string} key - The uri of a key stored in KMS
   * @param {Object} scr - Plaintext
   * @param {Object} options
   * @param {string} options.onBehalfOf - Fetch the KMS key on behalf of another user (using the user's UUID), active user requires the 'spark.kms_orgagent' role
   * @returns {string} Encrypted SCR
   */
  encryptScr: function encryptScr(key, scr, options) {
    /* istanbul ignore if */
    if (!scr.loc) {
      return _promise2.default.reject(new Error('Cannot encrypt `scr` without first setting `loc`'));
    }

    return this.getKey(key, options).then(function (k) {
      return scr.toJWE(k.jwk);
    });
  },


  /**
   * Encrypt plaintext using the supplied key uri.
   *
   * @param {string} key - The uri of a key stored in KMS
   * @param {string} plaintext
   * @param {Object} options
   * @param {string} options.onBehalfOf - Fetch the KMS key on behalf of another user (using the user's UUID), active user requires the 'spark.kms_orgagent' role
   * @returns {string} Encrypted text
   */
  encryptText: function encryptText(key, plaintext, options) {
    var _this3 = this;

    return this.getKey(key, options).then(function (k) {
      return _nodeJose2.default.JWE.createEncrypt(_this3.config.joseOptions, {
        key: k.jwk,
        header: {
          alg: 'dir'
        },
        reference: null
      }).final(plaintext, 'utf8');
    });
  },


  /**
   * Fetch the key associated with the supplied KMS uri.
   *
   * @param {string} uri - The uri of a key stored in KMS
   * @param {Object} options
   * @param {string} options.onBehalfOf - Fetch the KMS key on behalf of another user (using the user's UUID), active user requires the 'spark.kms_orgagent' role
   * @returns {string} Key
   */
  getKey: function getKey(uri) {
    var _this4 = this;

    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
        onBehalfOf = _ref.onBehalfOf;

    if (uri.jwk) {
      return this.kms.asKey(uri);
    }

    var storageKey = uri;

    if (onBehalfOf) {
      storageKey += '/onBehalfOf/' + onBehalfOf;
    }

    return this.unboundedStorage.get(storageKey).then(function (keyString) {
      return JSON.parse(keyString);
    }).then(function (keyObject) {
      return _this4.kms.asKey(keyObject);
    }).catch(function () {
      return _this4.kms.fetchKey({ uri: uri, onBehalfOf: onBehalfOf }).then((0, _common.tap)(function (key) {
        return _this4.unboundedStorage.put(storageKey, (0, _stringify2.default)(key, replacer));
      }));
    });
  },
  version: '1.72.6'
});

/**
 * JSON.stringify replacer that ensures private key data is serialized.
 * @param {string} k
 * @param {mixed} v
 * @returns {mixed}
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

function replacer(k, v) {
  if (k === 'jwk') {
    // note: this[k] and v may be different representations of the same value
    // eslint-disable-next-line no-invalid-this
    var json = this[k].toJSON(true);

    return json;
  }

  return v;
}

exports.default = Encryption;
