// @webex/internal-plugin-encryption/dist/kms-errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DryError = exports.KmsTimeoutError = exports.KmsError = undefined;

var _apply = require('babel-runtime/core-js/reflect/apply');

var _apply2 = _interopRequireDefault(_apply);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp, _class2, _temp2; /*!
                                     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                                     */

var _common = require('@webex/common');

var _webexCore = require('@webex/webex-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Error class for KMS errors
 */
var KmsError = exports.KmsError = (_temp = _class = function (_Exception) {
  (0, _inherits3.default)(KmsError, _Exception);

  function KmsError() {
    (0, _classCallCheck3.default)(this, KmsError);
    return (0, _possibleConstructorReturn3.default)(this, (KmsError.__proto__ || (0, _getPrototypeOf2.default)(KmsError)).apply(this, arguments));
  }

  (0, _createClass3.default)(KmsError, [{
    key: 'parse',


    /**
     * @param {HttpResponse} body
     * @returns {string}
     */
    value: function parse(body) {
      body = body.body || body;

      (0, _defineProperties2.default)(this, {
        body: {
          enumerable: false,
          value: body
        },
        reason: {
          enumerable: false,
          value: body.reason
        },
        requestId: {
          enumerable: false,
          value: body.requestId
        },
        status: {
          enumerable: false,
          value: body.status
        }
      });

      var message = typeof body === 'string' ? body : body.reason;

      if (!message) {
        message = this.constructor.defaultMessage;
      }
      if (body.status) {
        message += '\nKMS_RESPONSE_STATUS: ' + body.status;
      }
      if (body.requestId) {
        message += '\nKMS_REQUEST_ID: ' + body.requestId;
      }

      return message;
    }
  }]);
  return KmsError;
}(_common.Exception), _class.defaultMessage = 'An unknown error occurred while communicating with the kms. This implies we received an error response without a body.', _temp);

/**
 * Thrown when an expected KMSResponse is not received in a timely manner
 */

var KmsTimeoutError = exports.KmsTimeoutError = function (_KmsError) {
  (0, _inherits3.default)(KmsTimeoutError, _KmsError);

  function KmsTimeoutError() {
    (0, _classCallCheck3.default)(this, KmsTimeoutError);
    return (0, _possibleConstructorReturn3.default)(this, (KmsTimeoutError.__proto__ || (0, _getPrototypeOf2.default)(KmsTimeoutError)).apply(this, arguments));
  }

  (0, _createClass3.default)(KmsTimeoutError, [{
    key: 'parse',

    /**
     * @param {KmsRequest} options.request
     * @param {KmsRequest} options.timeout
     * @returns {string}
     */
    value: function parse() {
      var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
          _ref$request = _ref.request,
          request = _ref$request === undefined ? {} : _ref$request,
          timeout = _ref.timeout;

      var message = 'The KMS did not respond within ' + (timeout ? timeout + ' milliseconds' : 'a timely fashion');

      if (request) {
        if (request.method && request.uri) {
          message += '\nKMS_REQUEST: ' + request.method + ' ' + request.uri;
        }

        if (request.requestId) {
          message += '\nKMS_REQUEST_ID: ' + request.requestId;
        }
      }

      return message;
    }
  }]);
  return KmsTimeoutError;
}(KmsError);

/**
 * Emitted when a REST request includes an encrypter error
 */


var DryError = exports.DryError = (_temp2 = _class2 = function (_WebexHttpError) {
  (0, _inherits3.default)(DryError, _WebexHttpError);

  function DryError() {
    (0, _classCallCheck3.default)(this, DryError);
    return (0, _possibleConstructorReturn3.default)(this, (DryError.__proto__ || (0, _getPrototypeOf2.default)(DryError)).apply(this, arguments));
  }

  (0, _createClass3.default)(DryError, [{
    key: 'parse',


    /**
     * @param {WebexHttpError} reason
     * @returns {string}
     */
    value: function parse(reason) {
      (0, _apply2.default)(_webexCore.WebexHttpError.prototype.parse, this, [reason._res]);
      var body = reason._res.body.message;

      var message = body.reason || body;

      if (!message) {
        message = this.constructor.defaultMessage;
      }
      if (this.options.url) {
        message += '\n' + this.options.method + ' ' + this.options.url;
      } else if (this.options.uri) {
        message += '\n' + this.options.method + ' ' + this.options.uri;
      } else {
        message += '\n' + this.options.method + ' ' + this.options.service.toUpperCase() + '/' + this.options.resource;
      }
      message += '\nWEBEX_TRACKING_ID: ' + this.options.headers.trackingid;

      if (body.status) {
        message += '\nKMS_RESPONSE_STATUS: ' + body.status;
      }
      if (body.requestId) {
        message += '\nKMS_REQUEST_ID: ' + body.requestId;
      }

      (0, _defineProperties2.default)(this, {
        reason: {
          enumerable: false,
          value: body.reason
        },
        requestId: {
          enumerable: false,
          value: body.requestId
        },
        status: {
          enumerable: false,
          value: body.status
        }
      });

      return message;
    }
  }]);
  return DryError;
}(_webexCore.WebexHttpError), _class2.defaultMessage = 'An unknown error was received from a service that proxies to the KMS', _temp2);
