// @webex/internal-plugin-lyra/dist/lyra.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webexCore = require('@webex/webex-core');

var _space = require('./space');

var _space2 = _interopRequireDefault(_space);

var _device = require('./device');

var _device2 = _interopRequireDefault(_device);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Lyra = _webexCore.WebexPlugin.extend({
  /**
   * @typedef {Object} Endpoint
   * @property {Object} advertiser
   * @property {string} advertiser.id
   * @property {string} advertiser.displayName
   * @property {string} advertiser.orgId
   * @property {Object} links
   * @property {Object} links.addMeToSpace
   * @property {string} links.addMeToSpace.href
   * @property {string} links.addMeToSpace.method
   * @property {Object} links.lyra_space
   * @property {string} links.lyra_space.href
   * @property {string} links.lyra_space.method
   * @property {string} proof
   * @property {Object} token
   * @property {string} token.value
   */
  namespace: 'Lyra',
  children: {
    space: _space2.default,
    device: _device2.default
  },

  /**
   * Get the endpoint information
   * @param {string} token - ultrasound token decoded
   * @returns {Promise<Endpoint>}
   */
  getAdvertisedEndpoint: function getAdvertisedEndpoint(token) {
    return this.webex.request({
      method: 'GET',
      api: 'proximity',
      resource: '/ultrasound/advertisements',
      qs: {
        token: token
      }
    }).then(function (res) {
      return res.body;
    });
  },
  version: '1.72.6'
}); /*!
     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
     */

exports.default = Lyra;
