// @webex/internal-plugin-calendar/dist/calendar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webexCore = require('@webex/webex-core');

var Calendar = _webexCore.WebexPlugin.extend({
  namespace: 'Calendar',

  /**
   * Decrypts an encrypted incoming calendar event
   * @param {Object} event
   * @returns {Promise} Resolves with a decrypted calendar event
   */
  processMeetingEvent: function processMeetingEvent(event) {
    return this.webex.transform('inbound', event).then(function () {
      return event;
    });
  },


  /**
   * Retrieves a collection of meetings based on the request parameters
   * @param {Object} options
   * @param {Date} options.fromDate the start of the time range
   * @param {Date} options.toDate the end of the time range
   * @returns {Promise} Resolves with an array of meetings
   */
  list: function list(options) {
    options = options || {};

    return this.webex.request({
      method: 'GET',
      service: 'calendar',
      resource: 'calendarEvents',
      qs: options
    }).then(function (res) {
      return res.body.items;
    });
  },
  version: '1.72.6'
}); /*!
     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
     */

exports.default = Calendar;
