// @webex/internal-plugin-mercury/dist/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Forbidden = exports.NotAuthorized = exports.BadRequest = exports.UnknownResponse = exports.ConnectionError = undefined;

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp, _class2, _temp2, _class3, _temp3, _class4, _temp4, _class5, _temp5; /*!
                                                                                        * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                                                                                        */

var _common = require('@webex/common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Exception thrown when a websocket gets closed
 */
var ConnectionError = exports.ConnectionError = (_temp = _class = function (_Exception) {
  (0, _inherits3.default)(ConnectionError, _Exception);

  function ConnectionError() {
    (0, _classCallCheck3.default)(this, ConnectionError);
    return (0, _possibleConstructorReturn3.default)(this, (ConnectionError.__proto__ || (0, _getPrototypeOf2.default)(ConnectionError)).apply(this, arguments));
  }

  (0, _createClass3.default)(ConnectionError, [{
    key: 'parse',


    /**
     * @param {CloseEvent} event
     * @returns {string}
     */
    value: function parse() {
      var event = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      (0, _defineProperties2.default)(this, {
        code: {
          value: event.code
        },
        reason: {
          value: event.reason
        }
      });

      return event.reason;
    }
  }]);
  return ConnectionError;
}(_common.Exception), _class.defaultMessage = 'Failed to connect to socket', _temp);

/**
 * thrown for CloseCode 4400
 */

var UnknownResponse = exports.UnknownResponse = (_temp2 = _class2 = function (_ConnectionError) {
  (0, _inherits3.default)(UnknownResponse, _ConnectionError);

  function UnknownResponse() {
    (0, _classCallCheck3.default)(this, UnknownResponse);
    return (0, _possibleConstructorReturn3.default)(this, (UnknownResponse.__proto__ || (0, _getPrototypeOf2.default)(UnknownResponse)).apply(this, arguments));
  }

  return UnknownResponse;
}(ConnectionError), _class2.defaultMessage = 'UnknownResponse is produced by IE when we receive a 4XXX. You probably want to treat this like a NotFound', _temp2);

/**
 * thrown for CloseCode 4400
 */

var BadRequest = exports.BadRequest = (_temp3 = _class3 = function (_ConnectionError2) {
  (0, _inherits3.default)(BadRequest, _ConnectionError2);

  function BadRequest() {
    (0, _classCallCheck3.default)(this, BadRequest);
    return (0, _possibleConstructorReturn3.default)(this, (BadRequest.__proto__ || (0, _getPrototypeOf2.default)(BadRequest)).apply(this, arguments));
  }

  return BadRequest;
}(ConnectionError), _class3.defaultMessage = 'BadRequest usually implies an attempt to use service account credentials', _temp3);

/**
 * thrown for CloseCode 4401
 */

var NotAuthorized = exports.NotAuthorized = (_temp4 = _class4 = function (_ConnectionError3) {
  (0, _inherits3.default)(NotAuthorized, _ConnectionError3);

  function NotAuthorized() {
    (0, _classCallCheck3.default)(this, NotAuthorized);
    return (0, _possibleConstructorReturn3.default)(this, (NotAuthorized.__proto__ || (0, _getPrototypeOf2.default)(NotAuthorized)).apply(this, arguments));
  }

  return NotAuthorized;
}(ConnectionError), _class4.defaultMessage = 'Please refresh your access token', _temp4);

/**
 * thrown for CloseCode 4403
 */

var Forbidden = exports.Forbidden = (_temp5 = _class5 = function (_ConnectionError4) {
  (0, _inherits3.default)(Forbidden, _ConnectionError4);

  function Forbidden() {
    (0, _classCallCheck3.default)(this, Forbidden);
    return (0, _possibleConstructorReturn3.default)(this, (Forbidden.__proto__ || (0, _getPrototypeOf2.default)(Forbidden)).apply(this, arguments));
  }

  return Forbidden;
}(ConnectionError), _class5.defaultMessage = 'Forbidden usually implies these credentials are not entitled for Webex', _temp5);

// /**
//  * thrown for CloseCode 4404
//  */
// export class NotFound extends ConnectionError {
//   static defaultMessage = `Please refresh your Mercury registration (typically via a WDM refresh)`;
// }
