// @webex/internal-plugin-mercury/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnknownResponse = exports.NotAuthorized = exports.Forbidden = exports.ConnectionError = exports.BadRequest = exports.config = exports.Socket = exports.Mercury = exports.default = undefined;

var _mercury = require('./mercury');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mercury).default;
  }
});
Object.defineProperty(exports, 'Mercury', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mercury).default;
  }
});

var _socket = require('./socket');

Object.defineProperty(exports, 'Socket', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_socket).default;
  }
});

var _config = require('./config');

Object.defineProperty(exports, 'config', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_config).default;
  }
});

var _errors = require('./errors');

Object.defineProperty(exports, 'BadRequest', {
  enumerable: true,
  get: function get() {
    return _errors.BadRequest;
  }
});
Object.defineProperty(exports, 'ConnectionError', {
  enumerable: true,
  get: function get() {
    return _errors.ConnectionError;
  }
});
Object.defineProperty(exports, 'Forbidden', {
  enumerable: true,
  get: function get() {
    return _errors.Forbidden;
  }
});
Object.defineProperty(exports, 'NotAuthorized', {
  enumerable: true,
  get: function get() {
    return _errors.NotAuthorized;
  }
});
Object.defineProperty(exports, 'UnknownResponse', {
  enumerable: true,
  get: function get() {
    return _errors.UnknownResponse;
  }
});

require('@webex/internal-plugin-wdm');

require('@webex/internal-plugin-feature');

require('@webex/internal-plugin-metrics');

var _webexCore = require('@webex/webex-core');

var _mercury2 = _interopRequireDefault(_mercury);

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

(0, _webexCore.registerInternalPlugin)('mercury', _mercury2.default, {
  config: _config2.default,
  onBeforeLogout: function onBeforeLogout() {
    return this.disconnect();
  }
});
