// @webex/webex-core/dist/lib/credentials/credentials.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _apply = require('babel-runtime/core-js/reflect/apply');

var _apply2 = _interopRequireDefault(_apply);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _isEmpty2 = require('lodash/isEmpty');

var _isEmpty3 = _interopRequireDefault(_isEmpty2);

var _isObject2 = require('lodash/isObject');

var _isObject3 = _interopRequireDefault(_isObject2);

var _cloneDeep2 = require('lodash/cloneDeep');

var _cloneDeep3 = _interopRequireDefault(_cloneDeep2);

var _clone2 = require('lodash/clone');

var _clone3 = _interopRequireDefault(_clone2);

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _desc, _value, _obj; /*!
                                                                   * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                                                                   */

var _querystring = require('querystring');

var _querystring2 = _interopRequireDefault(_querystring);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _common = require('@webex/common');

var _commonTimers = require('@webex/common-timers');

var _webexPlugin = require('../webex-plugin');

var _webexPlugin2 = _interopRequireDefault(_webexPlugin);

var _decorators = require('../storage/decorators');

var _grantErrors = require('./grant-errors');

var _grantErrors2 = _interopRequireDefault(_grantErrors);

var _scope = require('./scope');

var _token = require('./token');

var _token2 = _interopRequireDefault(_token);

var _tokenCollection = require('./token-collection');

var _tokenCollection2 = _interopRequireDefault(_tokenCollection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

/**
 * @class
 */
var Credentials = _webexPlugin2.default.extend((_dec = (0, _common.oneFlight)({ keyFactory: function keyFactory(scope) {
    return scope;
  } }), _dec2 = (0, _decorators.waitForValue)('@'), _dec3 = (0, _decorators.persist)('@'), _dec4 = (0, _decorators.waitForValue)('@'), _dec5 = (0, _common.whileInFlight)('isRefreshing'), _dec6 = (0, _decorators.waitForValue)('@'), (_obj = {
  collections: {
    userTokens: _tokenCollection2.default
  },

  dataTypes: {
    token: (0, _common.makeStateDataType)(_token2.default, 'token').dataType
  },

  derived: {
    canAuthorize: {
      deps: ['supertoken', 'supertoken.canAuthorize', 'canRefresh'],
      fn: function fn() {
        return Boolean(this.supertoken && this.supertoken.canAuthorize || this.canRefresh);
      }
    },
    canRefresh: {
      deps: ['supertoken', 'supertoken.canRefresh'],
      fn: function fn() {
        // If we're operating in JWT mode, we have to delegate to the consumer
        if (this.config.jwtRefreshCallback) {
          return true;
        }

        return Boolean(this.supertoken && this.supertoken.canRefresh);
      }
    }
  },

  props: {
    supertoken: (0, _common.makeStateDataType)(_token2.default, 'token').prop
  },

  namespace: 'Credentials',

  session: {
    isRefreshing: {
      default: false,
      type: 'boolean'
    },
    /**
     * Becomes `true` once the {@link loaded} event fires.
     * @see {@link WebexPlugin#ready}
     * @instance
     * @memberof Credentials
     * @type {boolean}
     */
    ready: {
      default: false,
      type: 'boolean'
    },
    refreshTimer: {
      default: undefined,
      type: 'any'
    }
  },

  /**
   * Generates an OAuth Login URL. Prefers the api.ciscospark.com proxy if the
   * instance is initialize with an authorizatUrl, but fallsback to idbroker
   * as the base otherwise.
   * @instance
   * @memberof Credentials
   * @param {Object} [options={}]
   * @returns {string}
   */
  buildLoginUrl: function buildLoginUrl() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : { clientType: 'public' };

    /* eslint-disable camelcase */
    if (options.state && !(0, _isObject3.default)(options.state)) {
      throw new Error('if specified, `options.state` must be an object');
    }

    options.client_id = this.config.client_id;
    options.redirect_uri = this.config.redirect_uri;
    options.scope = this.config.scope;

    options = (0, _cloneDeep3.default)(options);

    if (!options.response_type) {
      options.response_type = options.clientType === 'public' ? 'token' : 'code';
    }
    (0, _deleteProperty2.default)(options, 'clientType');

    if (options.state) {
      if (!(0, _isEmpty3.default)(options.state)) {
        options.state = _common.base64.toBase64Url((0, _stringify2.default)(options.state));
      } else {
        delete options.state;
      }
    }

    return this.config.authorizeUrl + '?' + _querystring2.default.stringify(options);
    /* eslint-enable camelcase */
  },


  /**
   * Generates a Logout URL
   * @instance
   * @memberof Credentials
   * @param {Object} [options={}]
   * @returns {[type]}
   */
  buildLogoutUrl: function buildLogoutUrl() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    return this.config.logoutUrl + '?' + _querystring2.default.stringify((0, _assign2.default)({
      cisService: this.config.service,
      goto: this.config.redirect_uri
    }, options));
  },


  /**
   * Generates a number between 60% - 90% of expired value
   * @instance
   * @memberof Credentials
   * @param {number} expiration
   * @private
   * @returns {number}
   */
  calcRefreshTimeout: function calcRefreshTimeout(expiration) {
    return Math.floor((Math.floor(Math.random() * 4) + 6) / 10 * expiration);
  },
  constructor: function constructor() {
    var _this = this;

    // HACK to deal with the fact that AmpersandState#dataTypes#set is a pure
    // function.
    this._dataTypes = (0, _cloneDeep3.default)(this._dataTypes);
    (0, _keys2.default)(this._dataTypes).forEach(function (key) {
      if (_this._dataTypes[key].set) {
        _this._dataTypes[key].set = _this._dataTypes[key].set.bind(_this);
      }
    });
    // END HACK

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    (0, _apply2.default)(_webexPlugin2.default, this, args);
  },


  /**
   * Downscopes a token
   * @instance
   * @memberof Credentials
   * @param {string} scope
   * @private
   * @returns {Promise<Token>}
   */
  downscope: function downscope(scope) {
    var _this2 = this;

    return this.supertoken.downscope(scope).catch(function (reason) {
      _this2.logger.trace('credentials: failed to downscope supertoken to ' + scope, reason);
      _this2.logger.trace('credentials: falling back to supertoken for ' + scope);

      return _promise2.default.resolve(new _token2.default((0, _assign2.default)({ scope: scope }, _this2.supertoken.serialize())), { parent: _this2 });
    });
  },


  /**
   * Requests a client credentials grant and returns the token. Given the
   * limited use for such tokens as this time, this method does not cache its
   * token.
   * @instance
   * @memberof Credentials
   * @param {Object} options
   * @returns {Promise<Token>}
   */
  getClientToken: function getClientToken() {
    var _this3 = this;

    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    this.logger.info('credentials: requesting client credentials grant');

    options = options || {};
    options.scope = options.scope || 'webexsquare:admin';

    return this.webex.request({
      /* eslint-disable camelcase */
      method: 'POST',
      uri: this.config.tokenUrl,
      form: {
        grant_type: 'client_credentials',
        scope: options.scope,
        self_contained_token: true
      },
      auth: {
        user: this.config.client_id,
        pass: this.config.client_secret,
        sendImmediately: true
      },
      shouldRefreshAccessToken: false
      /* eslint-enable camelcase */
    }).then(function (res) {
      return new _token2.default(res.body, { parent: _this3 });
    }).catch(function (res) {
      if (res.statusCode !== 400) {
        return _promise2.default.reject(res);
      }

      var ErrorConstructor = _grantErrors2.default.select(res.body.error);

      return _promise2.default.reject(new ErrorConstructor(res._res || res));
    });
  },

  /**
   * Resolves with a token with the specified scopes. If no scope is specified,
   * defaults to omit(webex.credentials.scope, 'spark:kms'). If no such token is
   * available, downscopes the supertoken to that scope.
   * @instance
   * @memberof Credentials
   * @param {string} scope
   * @returns {Promise<Token>}
   */
  getUserToken: function getUserToken(scope) {
    var _this4 = this;

    return _promise2.default.resolve(!this.isRefreshing || new _promise2.default(function (resolve) {
      _this4.logger.info('credentials: token refresh inflight; delaying getUserToken until refresh completes');
      _this4.once('change:isRefreshing', function () {
        _this4.logger.info('credentials: token refresh complete; reinvoking getUserToken');
        resolve();
      });
    })).then(function () {
      if (!_this4.canAuthorize) {
        _this4.logger.info('credentials: cannot produce an access token from current state');

        return _promise2.default.reject(new Error('Current state cannot produce an access token'));
      }

      if (!scope) {
        scope = (0, _scope.filterScope)('spark:kms', _this4.config.scope);
      }

      scope = (0, _scope.sortScope)(scope);

      if (scope === (0, _scope.sortScope)(_this4.config.scope)) {
        return _promise2.default.resolve(_this4.supertoken);
      }

      var token = _this4.userTokens.get(scope);

      // we should also check for the token.access_token since token object does
      // not get cleared on unsetting while logging out.
      if (!token || !token.access_token) {
        return _this4.downscope(scope).then((0, _common.tap)(function (t) {
          return _this4.userTokens.add(t);
        }));
      }

      return _promise2.default.resolve(token);
    });
  },

  /**
   * Initializer
   * @instance
   * @memberof Credentials
   * @param {Object} attrs
   * @param {Object} options
   * @private
   * @returns {Credentials}
   */
  initialize: function initialize(attrs, options) {
    var _this5 = this;

    if (attrs) {
      if (typeof attrs === 'string') {
        this.supertoken = attrs;
      }

      if (attrs.access_token) {
        this.supertoken = attrs;
      }

      if (attrs.authorization) {
        if (attrs.authorization.supertoken) {
          this.supertoken = attrs.authorization.supertoken;
        } else {
          this.supertoken = attrs.authorization;
        }
      }

      // schedule refresh
      if (this.supertoken && this.supertoken.expires) {
        this.scheduleRefresh(this.supertoken.expires);
      }
    }

    (0, _apply2.default)(_webexPlugin2.default.prototype.initialize, this, [attrs, options]);

    this.listenToOnce(this.parent, 'change:config', function () {
      if (_this5.config.authorizationString) {
        var parsed = _url2.default.parse(_this5.config.authorizationString, true);

        /* eslint-disable camelcase */
        _this5.config.client_id = parsed.query.client_id;
        _this5.config.redirect_uri = parsed.query.redirect_uri;
        _this5.config.scope = parsed.query.scope;
        _this5.config.authorizeUrl = parsed.href.substr(0, parsed.href.indexOf('?'));
        /* eslint-enable camelcase */
      }
    });

    this.webex.once('loaded', function () {
      _this5.ready = true;
    });
  },

  /**
   * Clears all tokens from store them from the stores.
   *
   * This is no longer quite the right name for this method, but all of the
   * alternatives I'm coming up with are already taken.
   * @instance
   * @memberof Credentials
   * @returns {Promise}
   */
  invalidate: function invalidate() {
    this.logger.info('credentials: invalidating tokens');

    // clear refresh timer
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
      this.unset('refreshTimer');
    }

    try {
      this.unset('supertoken');
    } catch (err) {
      this.logger.warn('credentials: failed to clear supertoken', err);
    }

    while (this.userTokens.models.length) {
      try {
        this.userTokens.remove(this.userTokens.models[0]);
      } catch (err) {
        this.logger.warn('credentials: failed to remove user token', err);
      }
    }

    this.logger.info('credentials: finished removing tokens');

    // Return a promise to give the storage layer a tick or two to clear
    // localStorage
    return _promise2.default.resolve();
  },

  /**
   * Removes the supertoken and child tokens, then refreshes the supertoken;
   * subsequent calls to {@link Credentials#getUserToken()} will re-downscope
   * child tokens. Enqueus revocation of previous previousTokens. Yes, that's
   * the correct number of "previous"es.
   * @instance
   * @memberof Credentials
   * @returns {Promise}
   */
  refresh: function refresh() {
    var _this6 = this;

    this.logger.info('credentials: refresh requested');

    var supertoken = this.supertoken;

    var tokens = (0, _clone3.default)(this.userTokens.models);

    // This is kind of a leaky abstraction, since it relies on the authorization
    // plugin, but the only alternatives I see are
    // 1. put all JWT support in core
    // 2. have separate jwt and non-jwt auth plugins
    // while I like #2 from a code simplicity standpoint, the third-party DX
    // isn't great
    if (this.config.jwtRefreshCallback) {
      return this.config.jwtRefreshCallback(this.webex).then(function (jwt) {
        return _this6.webex.authorization.requestAccessTokenFromJwt({ jwt: jwt });
      });
    }

    return supertoken.refresh().then(function (st) {
      // clear refresh timer
      if (_this6.refreshTimer) {
        clearTimeout(_this6.refreshTimer);
        _this6.unset('refreshTimer');
      }
      _this6.supertoken = st;

      return _promise2.default.all(tokens.map(function (token) {
        return _this6.downscope(token.scope)
        // eslint-disable-next-line max-nested-callbacks
        .then(function (t) {
          _this6.logger.info('credentials: revoking token for ' + token.scope);

          return token.revoke().catch(function (err) {
            _this6.logger.warn('credentials: failed to revoke user token', err);
          }).then(function () {
            _this6.userTokens.remove(token.scope);
            _this6.userTokens.add(t);
          });
        });
      }));
    }).then(function () {
      _this6.scheduleRefresh(_this6.supertoken.expires);
    }).catch(function (error) {
      var InvalidRequestError = _grantErrors2.default.InvalidRequestError;


      if (error instanceof InvalidRequestError) {
        // Error: The refresh token provided is expired, revoked, malformed, or invalid. Hence emit an event to the client, an opportunity to logout.
        _this6.unset('supertoken');
        while (_this6.userTokens.models.length) {
          try {
            _this6.userTokens.remove(_this6.userTokens.models[0]);
          } catch (err) {
            _this6.logger.warn('credentials: failed to remove user token', err);
          }
        }
        _this6.webex.trigger('client:InvalidRequestError');
      }

      return _promise2.default.reject(error);
    });
  },


  /**
   * Schedules a token refresh or refreshes the token if token has expired
   * @instance
   * @memberof Credentials
   * @param {number} expires
   * @private
   * @returns {undefined}
   */
  scheduleRefresh: function scheduleRefresh(expires) {
    var _this7 = this;

    var expiresIn = expires - Date.now();

    if (expiresIn > 0) {
      var timeoutLength = this.calcRefreshTimeout(expiresIn);

      this.refreshTimer = (0, _commonTimers.safeSetTimeout)(function () {
        return _this7.refresh();
      }, timeoutLength);
    } else {
      this.refresh();
    }
  },
  version: '1.72.6'
}, (_applyDecoratedDescriptor(_obj, 'getUserToken', [_dec, _dec2], (0, _getOwnPropertyDescriptor2.default)(_obj, 'getUserToken'), _obj), _applyDecoratedDescriptor(_obj, 'initialize', [_dec3], (0, _getOwnPropertyDescriptor2.default)(_obj, 'initialize'), _obj), _applyDecoratedDescriptor(_obj, 'invalidate', [_common.oneFlight, _dec4], (0, _getOwnPropertyDescriptor2.default)(_obj, 'invalidate'), _obj), _applyDecoratedDescriptor(_obj, 'refresh', [_common.oneFlight, _dec5, _dec6], (0, _getOwnPropertyDescriptor2.default)(_obj, 'refresh'), _obj)), _obj)));

exports.default = Credentials;
