// @webex/webex-core/dist/lib/credentials/grant-errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OAuthError = undefined;

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _common = require('@webex/common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Error thrown during oauth flow
 */
var OAuthError = exports.OAuthError = function (_Exception) {
  (0, _inherits3.default)(OAuthError, _Exception);

  function OAuthError() {
    (0, _classCallCheck3.default)(this, OAuthError);
    return (0, _possibleConstructorReturn3.default)(this, (OAuthError.__proto__ || (0, _getPrototypeOf2.default)(OAuthError)).apply(this, arguments));
  }

  (0, _createClass3.default)(OAuthError, [{
    key: 'parse',

    /**
     * @param {HttpResponse} res
     * @returns {string}
     */
    value: function parse(res) {
      // This error may be returned in an HTTPResponse or as part of a querystring
      var body = res.body || res;

      (0, _defineProperties2.default)(this, {
        error: {
          enumerable: true,
          value: body.error
        },
        errorDescription: {
          enumerable: true,
          value: body.error_description
        },
        errorUri: {
          enumerable: true,
          value: body.error_uri
        },
        res: {
          enumerable: false,
          value: res
        }
      });

      return this.errorDescription;
    }
  }]);
  return OAuthError;
}(_common.Exception);
/**
 * InvalidRequestError
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

/* eslint camelcase: [0] */

var InvalidRequestError = function (_OAuthError) {
  (0, _inherits3.default)(InvalidRequestError, _OAuthError);

  function InvalidRequestError() {
    (0, _classCallCheck3.default)(this, InvalidRequestError);
    return (0, _possibleConstructorReturn3.default)(this, (InvalidRequestError.__proto__ || (0, _getPrototypeOf2.default)(InvalidRequestError)).apply(this, arguments));
  }

  return InvalidRequestError;
}(OAuthError);

/**
 * InvalidClientError
 */


var InvalidClientError = function (_OAuthError2) {
  (0, _inherits3.default)(InvalidClientError, _OAuthError2);

  function InvalidClientError() {
    (0, _classCallCheck3.default)(this, InvalidClientError);
    return (0, _possibleConstructorReturn3.default)(this, (InvalidClientError.__proto__ || (0, _getPrototypeOf2.default)(InvalidClientError)).apply(this, arguments));
  }

  return InvalidClientError;
}(OAuthError);

/**
 * InvalidGrantError
 */


var InvalidGrantError = function (_OAuthError3) {
  (0, _inherits3.default)(InvalidGrantError, _OAuthError3);

  function InvalidGrantError() {
    (0, _classCallCheck3.default)(this, InvalidGrantError);
    return (0, _possibleConstructorReturn3.default)(this, (InvalidGrantError.__proto__ || (0, _getPrototypeOf2.default)(InvalidGrantError)).apply(this, arguments));
  }

  return InvalidGrantError;
}(OAuthError);

/**
 * UnauthorizedClientError
 */


var UnauthorizedClientError = function (_OAuthError4) {
  (0, _inherits3.default)(UnauthorizedClientError, _OAuthError4);

  function UnauthorizedClientError() {
    (0, _classCallCheck3.default)(this, UnauthorizedClientError);
    return (0, _possibleConstructorReturn3.default)(this, (UnauthorizedClientError.__proto__ || (0, _getPrototypeOf2.default)(UnauthorizedClientError)).apply(this, arguments));
  }

  return UnauthorizedClientError;
}(OAuthError);

/**
 * UnsupportGrantTypeError
 */


var UnsupportGrantTypeError = function (_OAuthError5) {
  (0, _inherits3.default)(UnsupportGrantTypeError, _OAuthError5);

  function UnsupportGrantTypeError() {
    (0, _classCallCheck3.default)(this, UnsupportGrantTypeError);
    return (0, _possibleConstructorReturn3.default)(this, (UnsupportGrantTypeError.__proto__ || (0, _getPrototypeOf2.default)(UnsupportGrantTypeError)).apply(this, arguments));
  }

  return UnsupportGrantTypeError;
}(OAuthError);

/**
 * InvalidScopeError
 */


var InvalidScopeError = function (_OAuthError6) {
  (0, _inherits3.default)(InvalidScopeError, _OAuthError6);

  function InvalidScopeError() {
    (0, _classCallCheck3.default)(this, InvalidScopeError);
    return (0, _possibleConstructorReturn3.default)(this, (InvalidScopeError.__proto__ || (0, _getPrototypeOf2.default)(InvalidScopeError)).apply(this, arguments));
  }

  return InvalidScopeError;
}(OAuthError);

var errors = {
  OAuthError: OAuthError,
  InvalidRequestError: InvalidRequestError,
  InvalidClientError: InvalidClientError,
  InvalidGrantError: InvalidGrantError,
  UnauthorizedClientError: UnauthorizedClientError,
  UnsupportGrantTypeError: UnsupportGrantTypeError,
  InvalidScopeError: InvalidScopeError,
  invalid_request: InvalidRequestError,
  invalid_client: InvalidClientError,
  invalid_grant: InvalidGrantError,
  unauthorized_client: UnauthorizedClientError,
  unsupported_grant_type: UnsupportGrantTypeError,
  invalid_scope: InvalidScopeError,
  select: function select(errorString) {
    return errors[errorString] || OAuthError;
  }
};

exports.default = errors;
