// @webex/webex-core/dist/lib/credentials/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.grantErrors = exports.Token = exports.sortScope = exports.filterScope = exports.Credentials = undefined;

var _credentials = require('./credentials');

Object.defineProperty(exports, 'Credentials', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_credentials).default;
  }
});

var _scope = require('./scope');

Object.defineProperty(exports, 'filterScope', {
  enumerable: true,
  get: function get() {
    return _scope.filterScope;
  }
});
Object.defineProperty(exports, 'sortScope', {
  enumerable: true,
  get: function get() {
    return _scope.sortScope;
  }
});

var _token = require('./token');

Object.defineProperty(exports, 'Token', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_token).default;
  }
});

var _grantErrors = require('./grant-errors');

Object.defineProperty(exports, 'grantErrors', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_grantErrors).default;
  }
});

var _credentials2 = _interopRequireDefault(_credentials);

var _webexCore = require('../../webex-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

(0, _webexCore.registerPlugin)('credentials', _credentials2.default, {
  proxies: ['canAuthorize', 'canRefresh']
});
