// @webex/internal-plugin-wdm/dist/device/service-collection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _ampersandCollection = require('ampersand-collection');

var _ampersandCollection2 = _interopRequireDefault(_ampersandCollection);

var _serviceModel = require('./service-model');

var _serviceModel2 = _interopRequireDefault(_serviceModel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Collection of catalog services parsed from wdm registration.serviceHostMap.serviceLinks and
 * registration.serviceHostMap.hostCatalog
 * @class
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

var ServiceCollection = _ampersandCollection2.default.extend({
  mainIndex: 'service',
  model: _serviceModel2.default,

  /**
   * Mark the current host as failed and grab another url for connection
   * @param {string} uri Mark the host of this url as fail
   * @returns {string} new Url for connection
   */
  markFailedAndCycleUrl: function markFailedAndCycleUrl(uri) {
    if (!uri) {
      return _promise2.default.reject(new Error('`uri` is a required parameter'));
    }

    return this.inferServiceFromUrl(uri).then(function (service) {
      service.markHostFailed(uri);

      return service.cycleNextHost().then(function () {
        return service.url;
      });
    });
  },


  /**
   * Reset the available hosts if we are done trying all the host URLs
   * @param {string} uri to fetch the available hosts
   * @returns {string} new Url for connection
   */
  resetAllAndRetry: function resetAllAndRetry(uri) {
    if (!uri) {
      return _promise2.default.reject(new Error('`uri` is a required parameter'));
    }

    return this.inferServiceFromUrl(uri).then(function (service) {
      service.resetAllHosts();

      return service.url;
    });
  },


  /**
   * Find out what service this url belongs to (by looking at the host name)
   * @param {string} uri
   * @returns {Promise<ServiceModel>}
   */
  inferServiceFromUrl: function inferServiceFromUrl(uri) {
    var services = this.filter(function (service) {
      return service.doesUrlBelongToService(uri);
    });

    if (services.length >= 1) {
      return _promise2.default.resolve(services[0]);
    }

    return _promise2.default.reject(new Error('Unable to infer service for this url ' + uri));
  },


  /**
   * Find out what service this url belongs to, this returns the service name
   * instead of the service object
   * @param {string} uri Mark the host of this url as fail
   * @returns {Promise<ServiceModel.Service>}
   */
  inferServiceNameFromUrl: function inferServiceNameFromUrl(uri) {
    return this.inferServiceFromUrl(uri).then(function (service) {
      return service.service;
    });
  }
});

exports.default = ServiceCollection;
