// @webex/internal-plugin-wdm/dist/device/service-model.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _apply = require('babel-runtime/core-js/reflect/apply');

var _apply2 = _interopRequireDefault(_apply);

var _find2 = require('lodash/find');

var _find3 = _interopRequireDefault(_find2);

var _some2 = require('lodash/some');

var _some3 = _interopRequireDefault(_some2);

var _isObject2 = require('lodash/isObject');

var _isObject3 = _interopRequireDefault(_isObject2);

var _defaults2 = require('lodash/defaults');

var _defaults3 = _interopRequireDefault(_defaults2);

var _ampersandState = require('ampersand-state');

var _ampersandState2 = _interopRequireDefault(_ampersandState);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Represent a service parsed from wdm registration.serviceHostMap.serviceLinks and
 * registration.serviceHostMap.hostCatalog
 * @param {string} service service name from key of registration.serviceHostMap.serviceLinks
 * @param {string} defaultUrl Url provided in registration.serviceHostMap.serviceLinks
 * @param {Array<Host>} availableHosts Available datacenters from
 * registration.serviceHostMap.hostCatalog sorted by priority
 * @param {string} url Basically is the defaultUrl replaced with host provided in
 * the catalog
 * @class
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

var ServiceModel = _ampersandState2.default.extend({
  /**
    * @typedef {Object} Host - Represent a datacenter
    * @property {int} priority - Closer to 0 is higher priority.
    * @property {string} host - Host name.
    * @property {boolean} failed - True when cannot connect to url.
    */

  props: {
    service: 'string',
    defaultUrl: 'string',
    availableHosts: {
      type: 'array',
      default: function _default() {
        return [];
      }
    }
  },

  session: {
    currentHostIndex: {
      type: 'number',
      default: 0
    }
  },

  derived: {
    url: {
      deps: ['defaultUrl', 'availableHosts', 'currentHostIndex'],
      fn: function fn() {
        if (this.availableHosts.length === 0) {
          return this.defaultUrl;
        }
        var host = void 0;

        if (this.currentHostIndex >= this.availableHosts.length) {
          host = this.availableHosts[this.availableHosts.length - 1];
        } else {
          host = this.availableHosts[this.currentHostIndex];
        }

        return this._changeUrlHost(this.defaultUrl, host.host);
      }
    }
  },

  constructor: function constructor(attrs, options) {
    options = options || {};
    (0, _defaults3.default)(options, { parse: true });

    return (0, _apply2.default)(_ampersandState2.default.prototype.constructor, this, [attrs, options]);
  },


  idAttribute: 'service',

  // Override AmpersandState.serialize so we can return the latest url
  serialize: function serialize() {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var attrs = (0, _apply2.default)(_ampersandState2.default.prototype.serialize, this, args);

    attrs.url = this.url;

    return attrs;
  },


  // Override parse
  parse: function parse(attrs) {
    if (!attrs) {
      return {};
    }

    if (attrs.availableHosts) {
      // ensure highest priority is at the top
      // using number value here instead boolean for IE and Edge
      // https://github.com/tc39/ecma262/issues/902
      attrs.availableHosts.sort(function (a, b) {
        return a.priority - b.priority;
      });
    }

    return attrs;
  },


  // Override set to make sure we always run parse()
  // See https://github.com/AmpersandJS/ampersand-state/issues/146 for related
  // bug
  set: function set(key, value, options) {
    var attrs = void 0;

    // Handle both `"key", value` and `{key: value}` -style arguments.
    // The next block is a direct copy from ampersand-state, so no need to test
    // both scenarios.
    /* istanbul ignore next */
    if ((0, _isObject3.default)(key) || key === null) {
      attrs = key;
      options = value;
    } else {
      attrs = {};
      attrs[key] = value;
    }

    attrs = this.parse(attrs, options);

    return (0, _apply2.default)(_ampersandState2.default.prototype.set, this, [attrs, options]);
  },


  /**
   * Mark the current host as failing or if a uri is provided, find the host
   * and mark it as fail
   * @param {string} uri Mark the host of this url as fail
   * @returns {undefined}
   */
  markHostFailed: function markHostFailed(uri) {
    var host = this.getCurrentHost();

    if (uri) {
      var urlObj = _url2.default.parse(uri);

      host = (0, _find3.default)(this.availableHosts, function (h) {
        return h.host === urlObj.host;
      });
    }

    if (host) {
      host.failed = true;
    }
  },


  /**
   * Return the next available host, which is usually the next higher priority
   * host that has not yet been marked as failed
   * @returns {Promise<Host>}
   */
  cycleNextHost: function cycleNextHost() {
    for (var i = 0; i < this.availableHosts.length; i += 1) {
      var host = this.availableHosts[i];

      if (!host.failed && this.currentHostIndex !== i) {
        this.currentHostIndex = i;

        return _promise2.default.resolve(host);
      }
    }
    // this means all hosts have failed
    this.currentHostIndex = 0;

    return _promise2.default.reject(new Error('All hosts have failed for ' + this.service));
  },


  /**
   * Check if a url comes from this service
   * @param {string} uri
   * @returns {Boolean}
   */
  doesUrlBelongToService: function doesUrlBelongToService(uri) {
    var urlObj = _url2.default.parse(uri);
    var hosts = this.availableHosts.map(function (h) {
      return h.host;
    });

    hosts.push(_url2.default.parse(this.defaultUrl).host);

    return (0, _some3.default)(hosts, function (host) {
      return host === urlObj.host;
    });
  },


  /**
   * Return the current host/datacenter
   * @returns {Host}
   */
  getCurrentHost: function getCurrentHost() {
    return this.availableHosts[this.currentHostIndex];
  },


  /**
   * Resets all host/datacenter for a retry
   * @returns {undefined}
   */
  resetAllHosts: function resetAllHosts() {
    this.availableHosts.forEach(function (host) {
      host.failed = false;
    });
  },


  /**
   * Replace provided url by the current active host
   * @param {string} uri
   * @returns {string} uri
   */
  replaceUrlWithCurrentHost: function replaceUrlWithCurrentHost(uri) {
    return this._changeUrlHost(uri, this.getCurrentHost().host);
  },
  _changeUrlHost: function _changeUrlHost(currentUrl, host) {
    var urlObj = _url2.default.parse(currentUrl);

    urlObj.host = host;

    return _url2.default.format(urlObj);
  }
});

exports.default = ServiceModel;
