// @webex/http-core/dist/http-error-subtypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = makeSubTypes;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

/**
 * @param {Error} Base
 * @returns {undefined}
 */
function makeSubTypes(Base) {
  /**
   * NetworkOrCORSError
   */
  var NetworkOrCORSError = function (_Base) {
    (0, _inherits3.default)(NetworkOrCORSError, _Base);

    function NetworkOrCORSError() {
      (0, _classCallCheck3.default)(this, NetworkOrCORSError);
      return (0, _possibleConstructorReturn3.default)(this, (NetworkOrCORSError.__proto__ || (0, _getPrototypeOf2.default)(NetworkOrCORSError)).apply(this, arguments));
    }

    return NetworkOrCORSError;
  }(Base);
  /**
   * BadRequest
   */


  var BadRequest = function (_Base2) {
    (0, _inherits3.default)(BadRequest, _Base2);

    function BadRequest() {
      (0, _classCallCheck3.default)(this, BadRequest);
      return (0, _possibleConstructorReturn3.default)(this, (BadRequest.__proto__ || (0, _getPrototypeOf2.default)(BadRequest)).apply(this, arguments));
    }

    return BadRequest;
  }(Base);
  /**
   * Unauthorized
   */


  var Unauthorized = function (_BadRequest) {
    (0, _inherits3.default)(Unauthorized, _BadRequest);

    function Unauthorized() {
      (0, _classCallCheck3.default)(this, Unauthorized);
      return (0, _possibleConstructorReturn3.default)(this, (Unauthorized.__proto__ || (0, _getPrototypeOf2.default)(Unauthorized)).apply(this, arguments));
    }

    return Unauthorized;
  }(BadRequest);
  /**
   * PaymentRequired
   */


  var PaymentRequired = function (_BadRequest2) {
    (0, _inherits3.default)(PaymentRequired, _BadRequest2);

    function PaymentRequired() {
      (0, _classCallCheck3.default)(this, PaymentRequired);
      return (0, _possibleConstructorReturn3.default)(this, (PaymentRequired.__proto__ || (0, _getPrototypeOf2.default)(PaymentRequired)).apply(this, arguments));
    }

    return PaymentRequired;
  }(BadRequest);
  /**
   * Forbidden
   */


  var Forbidden = function (_BadRequest3) {
    (0, _inherits3.default)(Forbidden, _BadRequest3);

    function Forbidden() {
      (0, _classCallCheck3.default)(this, Forbidden);
      return (0, _possibleConstructorReturn3.default)(this, (Forbidden.__proto__ || (0, _getPrototypeOf2.default)(Forbidden)).apply(this, arguments));
    }

    return Forbidden;
  }(BadRequest);
  /**
   * NotFound
   */


  var NotFound = function (_BadRequest4) {
    (0, _inherits3.default)(NotFound, _BadRequest4);

    function NotFound() {
      (0, _classCallCheck3.default)(this, NotFound);
      return (0, _possibleConstructorReturn3.default)(this, (NotFound.__proto__ || (0, _getPrototypeOf2.default)(NotFound)).apply(this, arguments));
    }

    return NotFound;
  }(BadRequest);
  /**
   * MethodNotAllowed
   */


  var MethodNotAllowed = function (_BadRequest5) {
    (0, _inherits3.default)(MethodNotAllowed, _BadRequest5);

    function MethodNotAllowed() {
      (0, _classCallCheck3.default)(this, MethodNotAllowed);
      return (0, _possibleConstructorReturn3.default)(this, (MethodNotAllowed.__proto__ || (0, _getPrototypeOf2.default)(MethodNotAllowed)).apply(this, arguments));
    }

    return MethodNotAllowed;
  }(BadRequest);
  /**
   * NotAcceptable
   */


  var NotAcceptable = function (_BadRequest6) {
    (0, _inherits3.default)(NotAcceptable, _BadRequest6);

    function NotAcceptable() {
      (0, _classCallCheck3.default)(this, NotAcceptable);
      return (0, _possibleConstructorReturn3.default)(this, (NotAcceptable.__proto__ || (0, _getPrototypeOf2.default)(NotAcceptable)).apply(this, arguments));
    }

    return NotAcceptable;
  }(BadRequest);
  /**
   * ProxyAuthenticationRequired
   */


  var ProxyAuthenticationRequired = function (_BadRequest7) {
    (0, _inherits3.default)(ProxyAuthenticationRequired, _BadRequest7);

    function ProxyAuthenticationRequired() {
      (0, _classCallCheck3.default)(this, ProxyAuthenticationRequired);
      return (0, _possibleConstructorReturn3.default)(this, (ProxyAuthenticationRequired.__proto__ || (0, _getPrototypeOf2.default)(ProxyAuthenticationRequired)).apply(this, arguments));
    }

    return ProxyAuthenticationRequired;
  }(BadRequest);
  /**
   * RequestTimeout
   */


  var RequestTimeout = function (_BadRequest8) {
    (0, _inherits3.default)(RequestTimeout, _BadRequest8);

    function RequestTimeout() {
      (0, _classCallCheck3.default)(this, RequestTimeout);
      return (0, _possibleConstructorReturn3.default)(this, (RequestTimeout.__proto__ || (0, _getPrototypeOf2.default)(RequestTimeout)).apply(this, arguments));
    }

    return RequestTimeout;
  }(BadRequest);
  /**
   * Conflict
   */


  var Conflict = function (_BadRequest9) {
    (0, _inherits3.default)(Conflict, _BadRequest9);

    function Conflict() {
      (0, _classCallCheck3.default)(this, Conflict);
      return (0, _possibleConstructorReturn3.default)(this, (Conflict.__proto__ || (0, _getPrototypeOf2.default)(Conflict)).apply(this, arguments));
    }

    return Conflict;
  }(BadRequest);
  /**
   * Gone
   */


  var Gone = function (_BadRequest10) {
    (0, _inherits3.default)(Gone, _BadRequest10);

    function Gone() {
      (0, _classCallCheck3.default)(this, Gone);
      return (0, _possibleConstructorReturn3.default)(this, (Gone.__proto__ || (0, _getPrototypeOf2.default)(Gone)).apply(this, arguments));
    }

    return Gone;
  }(BadRequest);
  /**
   * LengthRequired
   */


  var LengthRequired = function (_BadRequest11) {
    (0, _inherits3.default)(LengthRequired, _BadRequest11);

    function LengthRequired() {
      (0, _classCallCheck3.default)(this, LengthRequired);
      return (0, _possibleConstructorReturn3.default)(this, (LengthRequired.__proto__ || (0, _getPrototypeOf2.default)(LengthRequired)).apply(this, arguments));
    }

    return LengthRequired;
  }(BadRequest);
  /**
   * PreconditionFailed
   */


  var PreconditionFailed = function (_BadRequest12) {
    (0, _inherits3.default)(PreconditionFailed, _BadRequest12);

    function PreconditionFailed() {
      (0, _classCallCheck3.default)(this, PreconditionFailed);
      return (0, _possibleConstructorReturn3.default)(this, (PreconditionFailed.__proto__ || (0, _getPrototypeOf2.default)(PreconditionFailed)).apply(this, arguments));
    }

    return PreconditionFailed;
  }(BadRequest);
  /**
   * RequestEntityTooLarge
   */


  var RequestEntityTooLarge = function (_BadRequest13) {
    (0, _inherits3.default)(RequestEntityTooLarge, _BadRequest13);

    function RequestEntityTooLarge() {
      (0, _classCallCheck3.default)(this, RequestEntityTooLarge);
      return (0, _possibleConstructorReturn3.default)(this, (RequestEntityTooLarge.__proto__ || (0, _getPrototypeOf2.default)(RequestEntityTooLarge)).apply(this, arguments));
    }

    return RequestEntityTooLarge;
  }(BadRequest);
  /**
   * RequestUriTooLong
   */


  var RequestUriTooLong = function (_BadRequest14) {
    (0, _inherits3.default)(RequestUriTooLong, _BadRequest14);

    function RequestUriTooLong() {
      (0, _classCallCheck3.default)(this, RequestUriTooLong);
      return (0, _possibleConstructorReturn3.default)(this, (RequestUriTooLong.__proto__ || (0, _getPrototypeOf2.default)(RequestUriTooLong)).apply(this, arguments));
    }

    return RequestUriTooLong;
  }(BadRequest);
  /**
   * UnsupportedMediaType
   */


  var UnsupportedMediaType = function (_BadRequest15) {
    (0, _inherits3.default)(UnsupportedMediaType, _BadRequest15);

    function UnsupportedMediaType() {
      (0, _classCallCheck3.default)(this, UnsupportedMediaType);
      return (0, _possibleConstructorReturn3.default)(this, (UnsupportedMediaType.__proto__ || (0, _getPrototypeOf2.default)(UnsupportedMediaType)).apply(this, arguments));
    }

    return UnsupportedMediaType;
  }(BadRequest);
  /**
   * RequestRangeNotSatisfiable
   */


  var RequestRangeNotSatisfiable = function (_BadRequest16) {
    (0, _inherits3.default)(RequestRangeNotSatisfiable, _BadRequest16);

    function RequestRangeNotSatisfiable() {
      (0, _classCallCheck3.default)(this, RequestRangeNotSatisfiable);
      return (0, _possibleConstructorReturn3.default)(this, (RequestRangeNotSatisfiable.__proto__ || (0, _getPrototypeOf2.default)(RequestRangeNotSatisfiable)).apply(this, arguments));
    }

    return RequestRangeNotSatisfiable;
  }(BadRequest);
  /**
   * ExpectationFailed
   */


  var ExpectationFailed = function (_BadRequest17) {
    (0, _inherits3.default)(ExpectationFailed, _BadRequest17);

    function ExpectationFailed() {
      (0, _classCallCheck3.default)(this, ExpectationFailed);
      return (0, _possibleConstructorReturn3.default)(this, (ExpectationFailed.__proto__ || (0, _getPrototypeOf2.default)(ExpectationFailed)).apply(this, arguments));
    }

    return ExpectationFailed;
  }(BadRequest);
  /**
   * TooManyRequests
   */


  var TooManyRequests = function (_BadRequest18) {
    (0, _inherits3.default)(TooManyRequests, _BadRequest18);

    function TooManyRequests() {
      (0, _classCallCheck3.default)(this, TooManyRequests);
      return (0, _possibleConstructorReturn3.default)(this, (TooManyRequests.__proto__ || (0, _getPrototypeOf2.default)(TooManyRequests)).apply(this, arguments));
    }

    return TooManyRequests;
  }(BadRequest);
  /**
   * InternalServerError
   */


  var InternalServerError = function (_Base3) {
    (0, _inherits3.default)(InternalServerError, _Base3);

    function InternalServerError() {
      (0, _classCallCheck3.default)(this, InternalServerError);
      return (0, _possibleConstructorReturn3.default)(this, (InternalServerError.__proto__ || (0, _getPrototypeOf2.default)(InternalServerError)).apply(this, arguments));
    }

    return InternalServerError;
  }(Base);
  /**
   * NotImplemented
   */


  var NotImplemented = function (_InternalServerError) {
    (0, _inherits3.default)(NotImplemented, _InternalServerError);

    function NotImplemented() {
      (0, _classCallCheck3.default)(this, NotImplemented);
      return (0, _possibleConstructorReturn3.default)(this, (NotImplemented.__proto__ || (0, _getPrototypeOf2.default)(NotImplemented)).apply(this, arguments));
    }

    return NotImplemented;
  }(InternalServerError);
  /**
   * BadGateway
   */


  var BadGateway = function (_InternalServerError2) {
    (0, _inherits3.default)(BadGateway, _InternalServerError2);

    function BadGateway() {
      (0, _classCallCheck3.default)(this, BadGateway);
      return (0, _possibleConstructorReturn3.default)(this, (BadGateway.__proto__ || (0, _getPrototypeOf2.default)(BadGateway)).apply(this, arguments));
    }

    return BadGateway;
  }(InternalServerError);
  /**
   * ServiceUnavailable
   */


  var ServiceUnavailable = function (_InternalServerError3) {
    (0, _inherits3.default)(ServiceUnavailable, _InternalServerError3);

    function ServiceUnavailable() {
      (0, _classCallCheck3.default)(this, ServiceUnavailable);
      return (0, _possibleConstructorReturn3.default)(this, (ServiceUnavailable.__proto__ || (0, _getPrototypeOf2.default)(ServiceUnavailable)).apply(this, arguments));
    }

    return ServiceUnavailable;
  }(InternalServerError);
  /**
   * GatewayTimeout
   */


  var GatewayTimeout = function (_InternalServerError4) {
    (0, _inherits3.default)(GatewayTimeout, _InternalServerError4);

    function GatewayTimeout() {
      (0, _classCallCheck3.default)(this, GatewayTimeout);
      return (0, _possibleConstructorReturn3.default)(this, (GatewayTimeout.__proto__ || (0, _getPrototypeOf2.default)(GatewayTimeout)).apply(this, arguments));
    }

    return GatewayTimeout;
  }(InternalServerError);
  /**
   * HttpVersionNotSupported
   */


  var HttpVersionNotSupported = function (_InternalServerError5) {
    (0, _inherits3.default)(HttpVersionNotSupported, _InternalServerError5);

    function HttpVersionNotSupported() {
      (0, _classCallCheck3.default)(this, HttpVersionNotSupported);
      return (0, _possibleConstructorReturn3.default)(this, (HttpVersionNotSupported.__proto__ || (0, _getPrototypeOf2.default)(HttpVersionNotSupported)).apply(this, arguments));
    }

    return HttpVersionNotSupported;
  }(InternalServerError);

  (0, _assign2.default)(Base, {
    0: NetworkOrCORSError,
    NetworkOrCORSError: NetworkOrCORSError,
    400: BadRequest,
    BadRequest: BadRequest,
    401: Unauthorized,
    Unauthorized: Unauthorized,
    402: PaymentRequired,
    PaymentRequired: PaymentRequired,
    403: Forbidden,
    Forbidden: Forbidden,
    404: NotFound,
    NotFound: NotFound,
    405: MethodNotAllowed,
    MethodNotAllowed: MethodNotAllowed,
    406: NotAcceptable,
    NotAcceptable: NotAcceptable,
    407: ProxyAuthenticationRequired,
    ProxyAuthenticationRequired: ProxyAuthenticationRequired,
    408: RequestTimeout,
    RequestTimeout: RequestTimeout,
    409: Conflict,
    Conflict: Conflict,
    410: Gone,
    Gone: Gone,
    411: LengthRequired,
    LengthRequired: LengthRequired,
    412: PreconditionFailed,
    PreconditionFailed: PreconditionFailed,
    413: RequestEntityTooLarge,
    RequestEntityTooLarge: RequestEntityTooLarge,
    414: RequestUriTooLong,
    RequestUriTooLong: RequestUriTooLong,
    415: UnsupportedMediaType,
    UnsupportedMediaType: UnsupportedMediaType,
    416: RequestRangeNotSatisfiable,
    RequestRangeNotSatisfiable: RequestRangeNotSatisfiable,
    417: ExpectationFailed,
    ExpectationFailed: ExpectationFailed,
    429: TooManyRequests,
    TooManyRequests: TooManyRequests,
    500: InternalServerError,
    InternalServerError: InternalServerError,
    501: NotImplemented,
    NotImplemented: NotImplemented,
    502: BadGateway,
    BadGateway: BadGateway,
    503: ServiceUnavailable,
    ServiceUnavailable: ServiceUnavailable,
    504: GatewayTimeout,
    GatewayTimeout: GatewayTimeout,
    505: HttpVersionNotSupported,
    HttpVersionNotSupported: HttpVersionNotSupported,
    select: function select(statusCode) {
      if (typeof statusCode === 'undefined' || statusCode === null) {
        return Base;
      }

      statusCode = statusCode.statusCode || statusCode;
      var E = Base[statusCode];

      if (E) {
        return E;
      }

      // Fallback to the default for the category (e.g. BadRequest for 429)
      statusCode = statusCode.toString().split('').shift() + '00';
      statusCode = parseInt(statusCode, 10);

      return Base[statusCode] || Base;
    }
  });
}
