// @webex/http-core/dist/http-error.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _class, _temp; /*!
                    * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                    */

var _common = require('@webex/common');

var _httpErrorSubtypes = require('./http-error-subtypes');

var _httpErrorSubtypes2 = _interopRequireDefault(_httpErrorSubtypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Base HttpError class. Unlikely to be thrown directly, but handy for general
 * type comparison.
 */
var HttpError = (_temp = _class = function (_Exception) {
  (0, _inherits3.default)(HttpError, _Exception);

  function HttpError() {
    (0, _classCallCheck3.default)(this, HttpError);
    return (0, _possibleConstructorReturn3.default)(this, (HttpError.__proto__ || (0, _getPrototypeOf2.default)(HttpError)).apply(this, arguments));
  }

  (0, _createClass3.default)(HttpError, [{
    key: 'parse',


    /**
     * Parses HttpResponse objects for useful information (status, headers, etc)
     * as well as attempting to extract a useful error message.
     * @param {HttpResponse} res
     * @returns {string}
     */

    /**
     * @example Extend via
     * MyError extends HttpError {
     *  static errorKeys = HttpError.errorKeys.concat([
     *    `key1`,
     *    `key2`
     *  ])
     * }
     *
     * @type {Array}
     */
    value: function parse(res) {
      // complexity is high here because of all the default values below.
      /* eslint complexity: [0] */
      var body = res.body;

      var message = void 0;

      switch (typeof body === 'undefined' ? 'undefined' : (0, _typeof3.default)(body)) {
        case 'string':
          try {
            body = JSON.parse(body);
            message = this.parseObject(body);
          } catch (err) {
            message = body;
          }
          break;
        case 'object':
          message = this.parseObject(body);
          break;
        default:
      }

      if (!message) {
        message = this.defaultMessage;
      }

      (0, _defineProperties2.default)(this, {
        body: {
          enumerable: false,
          value: body
        },
        httpVersion: {
          enumerable: false,
          value: res.httpVersion
        },
        headers: {
          enumerable: false,
          value: res.headers || {}
        },
        rawHeaders: {
          enumerable: false,
          value: res.rawHeaders || []
        },
        trailers: {
          enumerable: false,
          value: res.trailers || {}
        },
        rawTrailers: {
          enumerable: false,
          value: res.rawTrailers || []
        },
        method: {
          enumerable: false,
          value: res.method
        },
        url: {
          enumerable: false,
          value: res.url
        },
        statusCode: {
          enumerable: false,
          value: res.statusCode
        },
        statusMessage: {
          enumerable: false,
          value: res.statusMessage
        },
        socket: {
          enumerable: false,
          value: res.socket
        },
        _res: {
          enumerable: false,
          value: res
        }
      });

      return message;
    }

    /**
     * Recursively parses an error body looking for a likely error candidate
     * @param {object} body
     * @returns {string}
     */


    /**
     * Default error string if no error can be extracted from the http response
     * @type {string}
     */

  }, {
    key: 'parseObject',
    value: function parseObject(body) {
      // Search body for common names of error strings
      var messages = (0, _values2.default)((0, _pick3.default)(body, HttpError.errorKeys));

      // If no error candidate was found, stringify the entire body
      if (messages.length === 0) {
        return (0, _stringify2.default)(body, null, 2);
      }

      // Assume the first key found was the error explanation
      var message = messages[0];

      // If the explanation is an object, recurse and try again
      if ((typeof message === 'undefined' ? 'undefined' : (0, _typeof3.default)(message)) === 'object') {
        return this.parseObject(message);
      }

      // Return the first key
      return message;
    }
  }]);
  return HttpError;
}(_common.Exception), _class.errorKeys = ['error', 'errorString', 'response', 'errorResponse', 'message', 'msg'], _class.defaultMessage = 'An error was received while trying to fulfill the request', _temp);
exports.default = HttpError;


(0, _httpErrorSubtypes2.default)(HttpError);
HttpError.makeSubTypes = _httpErrorSubtypes2.default;
