// @webex/plugin-meetings/dist/roap/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _isNan = require('babel-runtime/core-js/number/is-nan');

var _isNan2 = _interopRequireDefault(_isNan);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp, _initialiseProps;

var _webexCore = require('@webex/webex-core');

var _constants = require('../constants');

var _loggerProxy = require('../common/logs/logger-proxy');

var _loggerProxy2 = _interopRequireDefault(_loggerProxy);

var _util = require('../meeting/util');

var _util2 = _interopRequireDefault(_util);

var _handler = require('./handler');

var _handler2 = _interopRequireDefault(_handler);

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Roap options
 * @typedef {Object} RoapOptions
 * @property {String} sdp
 * @property {Meeting} meeting
 * @property {Number} roapSeq
 */

/**
  * @typedef {Object} SeqOptions
  * @property {String} correlationId
  * @property {String} mediaId
  * @property {Number} seq
  */

/**
 * @class Roap
 * @export
 * @private
 */
var Roap = (_temp = _class = function (_StatelessWebexPlugin) {
  (0, _inherits3.default)(Roap, _StatelessWebexPlugin);

  /**
   *
   * @param {Object} attrs
   * @param {Object} options
   */
  function Roap(attrs, options) {
    (0, _classCallCheck3.default)(this, Roap);

    /**
     * @instance
     * @type {Object}
     * @private
     * @memberof Roap
     */
    var _this = (0, _possibleConstructorReturn3.default)(this, (Roap.__proto__ || (0, _getPrototypeOf2.default)(Roap)).call(this, {}, options));

    _initialiseProps.call(_this);

    _this.attrs = attrs;
    /**
     * @instance
     * @type {Object}
     * @private
     * @memberof Roap
     */
    _this.options = options;
    /**
     * The Roap Process State Handler
     * @instance
     * @type {RoapHandler}
     * @private
     * @memberof Roap
     */
    _this.roapHandler = new _handler2.default(_this.attrs, _this.options, _this.sendRoapOK.bind(_this), _this.sendRoapAnswer.bind(_this));
    /**
     * The Roap Request Server Proxy Object
     * @instance
     * @type {RoapRequest}
     * @private
     * @memberof Roap
     */
    _this.roapRequest = new _request2.default({}, options);
    /**
     * The last roap offer sent to server and acked
     * @instance
     * @type {Object}
     * @private
     * @memberof Roap
     */
    _this.lastRoapOffer = {};
    return _this;
  }

  /**
   * Starts listening to mercury events for Roap messages
   * @param {object} data event object
   * @returns {Promise}
   * @private
   * @memberof Roap
   */


  (0, _createClass3.default)(Roap, [{
    key: 'roapEvent',
    value: function roapEvent(data) {
      var msg = data.message;
      var correlationId = data.correlationId;


      _loggerProxy2.default.logger.log('Roap->roapEvent#Received Roap Message [' + (0, _stringify2.default)(msg, null, 2) + ']');
      this.roapHandler.submit({
        type: _constants.ROAP.RECEIVE_ROAP_MSG,
        msg: msg,
        correlationId: correlationId
      });
    }

    /**
     *
     * @param {String} correlationId correlation id of a meeting
     * @param {Number} seq ROAP sequence number
     * @returns {Promise}
     * @private
     * @memberof Roap
     */

  }, {
    key: 'stop',
    value: function stop(correlationId, seq) {
      this.roapHandler.submit({
        type: _constants.ROAP.RECEIVE_CALL_LEAVE,
        seq: seq,
        correlationId: correlationId
      });

      return _promise2.default.resolve();
    }

    /**
     *
     * @param {SeqOptions} options
     * @returns {null}
     * @private
     * @memberof Roap
     */

  }, {
    key: 'sendRoapOK',
    value: function sendRoapOK(options) {
      var _this2 = this;

      return _promise2.default.resolve().then(function () {
        var meeting = _this2.webex.meetings.meetingCollection.get(options.correlationId);
        var roapMessage = {
          messageType: _constants.ROAP.ROAP_TYPES.OK,
          version: _constants.ROAP.ROAP_VERSION,
          seq: options.seq
        };

        _loggerProxy2.default.logger.log('Roap->sendRoapOK#ROAP OK sending with seq ' + options.seq);

        return _this2.roapRequest.sendRoap({
          roapMessage: roapMessage,
          locusSelfUrl: meeting.selfUrl,
          mediaId: options.mediaId,
          correlationId: options.correlationId,
          audioMuted: meeting.isAudioMuted(),
          videoMuted: meeting.isVideoMuted(),
          meetingId: meeting.id
        }).then(function () {
          _this2.roapHandler.submit({
            type: _constants.ROAP.SEND_ROAP_MSG,
            msg: roapMessage,
            correlationId: options.correlationId
          });
          _loggerProxy2.default.logger.log('Roap->sendRoapOK#ROAP OK sent with seq ' + options.seq);
          meeting.setRoapSeq(options.seq);
        });
      });
    }

    // eslint-disable-next-line no-warning-comments
    // TODO: try to merge sendRoapOk and roapAnswer
    /**
     * Sends a ROAP answer...
     * @param {SeqOptions} options
     * @param {Boolean} options.audioMuted
     * @param {Boolean} options.videoMuted
     * @returns {Promise}
     * @private
     * @memberof Roap
     */

  }, {
    key: 'sendRoapAnswer',
    value: function sendRoapAnswer(options) {
      var _this3 = this;

      var meeting = this.webex.meetings.meetingCollection.get(options.correlationId);
      var roapMessage = {
        messageType: _constants.ROAP.ROAP_TYPES.ANSWER,
        sdps: options.sdps,
        version: _constants.ROAP.ROAP_VERSION,
        seq: options.seq
      };

      return this.roapRequest.sendRoap({
        roapMessage: roapMessage,
        locusSelfUrl: meeting.selfUrl,
        mediaId: options.mediaId,
        correlationId: options.correlationId,
        audioMuted: options.audioMuted,
        videoMuted: options.videoMuted,
        meetingId: meeting.id
      }).then(function () {
        meeting.setRoapSeq(options.seq);
        _this3.roapHandler.submit({
          type: _constants.ROAP.SEND_ROAP_MSG,
          msg: roapMessage,
          correlationId: options.correlationId
        });
      });
    }

    /**
     * Sends a ROAP error...
     * @param {Object} session
     * @param {Object} locus
     * @param {String} errorType
     * @returns {Promise}
     * @private
     * @memberof Roap
     */

  }, {
    key: 'sendRoapError',
    value: function sendRoapError(session, locus, errorType) {
      var msg = {
        messageType: _constants.ROAP.ROAP_TYPES.ERROR,
        version: _constants.ROAP.ROAP_VERSION,
        errorType: errorType,
        seq: session.OFFER.seq
      };

      return this.roapRequest.sendRoap(msg, locus);
    }

    /**
     * sends a roap media request
     * @param {RoapOptions} options
     * @returns {Promise}
     * @private
     * @memberof Roap
     */

  }, {
    key: 'sendRoapMediaRequest',
    value: function sendRoapMediaRequest(options) {
      var _this4 = this;

      var meeting = options.meeting;

      var roapMessage = {
        messageType: _constants.ROAP.ROAP_TYPES.OFFER,
        sdps: [options.sdp],
        // sdps: [options.sdp],
        version: _constants.ROAP.ROAP_VERSION,
        seq: typeof options.roapSeq !== 'number' && (0, _isNan2.default)(parseFloat(options.roapSeq)) ? 0 : options.roapSeq + 1,
        tieBreaker: 4294967294 // Math.floor(Math.random() * (2 ** 32) - 1) // TODO: Handle the roap  conflict scenario
      };

      this.roapHandler.submit({
        type: _constants.ROAP.SEND_ROAP_MSG,
        msg: roapMessage,
        correlationId: meeting.correlationId
      });

      return this.roapRequest.sendRoap({
        roapMessage: roapMessage,
        correlationId: meeting.correlationId,
        locusSelfUrl: meeting.selfUrl,
        mediaId: meeting.mediaId,
        audioMuted: meeting.isAudioMuted(),
        videoMuted: meeting.isVideoMuted(),
        meetingId: meeting.id
      }).then(function (locus) {
        _this4.roapHandler.submit({
          type: _constants.ROAP.SEND_ROAP_MSG_SUCCESS,
          seq: roapMessage.seq,
          correlationId: meeting.correlationId
        });
        meeting.setRoapSeq(roapMessage.seq);

        // eslint-disable-next-line no-warning-comments
        // TODO: we need to attach peerconenction to locus not sure if we need to pass everything here
        return locus;
        // eslint-disable-next-line no-warning-comments
        // TODO: check where to update the sequence number
      });
    }

    /**
     * sends a roap media request
     * @param {RoapOptions} options
     * @returns {Promise}
     * @private
     * @memberof Roap
     */

  }]);
  return Roap;
}(_webexCore.StatelessWebexPlugin), _initialiseProps = function _initialiseProps() {
  var _this5 = this;

  this.sendRoapCallRequest = function (options) {
    var meeting = options.meeting;

    var roapMessage = {
      messageType: _constants.ROAP.ROAP_TYPES.OFFER,
      sdps: [options.sdp],
      version: _constants.ROAP.ROAP_VERSION,
      seq: typeof options.roapSeq !== 'number' && (0, _isNan2.default)(parseFloat(options.roapSeq)) ? 0 : options.roapSeq + 1,
      tieBreaker: 4294967294 // Math.floor(Math.random() * (2 ** 32) - 1) // TODO: Handle the roap  conflict scenario
    };

    _this5.roapHandler.submit({
      type: _constants.ROAP.SEND_ROAP_MSG,
      msg: roapMessage,
      correlationId: meeting.correlationId
    });

    var roapBody = {
      localMedias: [{
        localSdp: (0, _stringify2.default)(_this5.roapRequest.attachRechabilityData({
          roapMessage: roapMessage,
          // eslint-disable-next-line no-warning-comments
          // TODO: check whats the need for video and audiomute
          audioMuted: !!meeting.isAudioMuted(),
          videoMuted: !!meeting.isVideoMuted()
        }))
        // mediaId: meeting.mediaId
      }]
    };

    return _util2.default.joinMeetingOptions(meeting, { roapMessage: roapBody }).then(function (locus) {
      _this5.roapHandler.submit({
        type: _constants.ROAP.SEND_ROAP_MSG_SUCCESS,
        seq: roapMessage.seq,
        correlationId: meeting.correlationId
      });
      meeting.setRoapSeq(roapMessage.seq);

      // eslint-disable-next-line no-warning-comments
      // TODO: we need to attach peerconenction to locus not sure if we need to pass everything here
      return locus;
      // eslint-disable-next-line no-warning-comments
      // TODO: check where to update the sequence number
    });
  };
}, _temp);
exports.default = Roap;
