// @webex/plugin-meetings/dist/meeting/in-meeting-actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class InMeetingActions
 */
var InMeetingActions = function () {
  function InMeetingActions() {
    (0, _classCallCheck3.default)(this, InMeetingActions);
    this.namespace = _constants.MEETINGS;

    this.canLock = null;
    this.canUnlock = null;
    this.canAssignHost = null;
  }

  (0, _createClass3.default)(InMeetingActions, [{
    key: 'setCanLock',
    value: function setCanLock(lock) {
      this.canLock = lock;
    }
  }, {
    key: 'setCanUnlock',
    value: function setCanUnlock(unlock) {
      this.canUnlock = unlock;
    }
  }, {
    key: 'setCanAssignHost',
    value: function setCanAssignHost(host) {
      this.canAssignHost = host;
    }
  }, {
    key: 'getCanLock',
    value: function getCanLock() {
      return this.canLock;
    }
  }, {
    key: 'getCanUnlock',
    value: function getCanUnlock() {
      return this.canUnlock;
    }
  }, {
    key: 'getCanAssignHost',
    value: function getCanAssignHost() {
      return this.canAssignHost;
    }
  }]);
  return InMeetingActions;
}(); /*!
      * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
      */

exports.default = InMeetingActions;
