// @webex/plugin-people/dist/people-batcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _webexCore = require('@webex/webex-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 * @ignore
 */
var PersonUUIDRequestBatcher = _webexCore.Batcher.extend({
  namespace: 'People',

  /**
   * @instance
   * @memberof PersonUUIDRequestBatcher
   * @param {HttpResponseObject} res
   * @returns {Promise}
   */
  handleHttpSuccess: function handleHttpSuccess(res) {
    var _this = this;

    return _promise2.default.all(res.body.items.map(function (personResponse) {
      return _this.handleItemSuccess(personResponse.id, personResponse);
    }));
  },
  handleItemFailure: function handleItemFailure(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.reject(response);
    });
  },


  /**
   * @instance
   * @memberof PersonUUIDRequestBatcher
   * @param {string} email
   * @param {Object} response
   * @returns {Promise}
   */
  handleItemSuccess: function handleItemSuccess(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.resolve(response);
    });
  },


  /**
   * @instance
   * @memberof PersonUUIDRequestBatcher
   * @param {string} uuidOrHydraId
   * @returns {Promise<string>}
   */
  fingerprintRequest: function fingerprintRequest(uuidOrHydraId) {
    var hydraId = this.webex.people.inferPersonIdFromUuid(uuidOrHydraId);

    return _promise2.default.resolve(hydraId);
  },


  /**
   * @instance
   * @memberof PersonUUIDRequestBatcher
   * @param {string} hydraId
   * @returns {Promise<string>}
   */
  fingerprintResponse: function fingerprintResponse(hydraId) {
    return _promise2.default.resolve(hydraId);
  },
  prepareRequest: function prepareRequest(ids) {
    return _promise2.default.resolve(ids.join());
  },


  /**
   * @instance
   * @memberof PersonUUIDRequestBatcher
   * @param {Object} ids
   * @returns {Promise<HttpResponseObject>}
   */
  submitHttpRequest: function submitHttpRequest(ids) {
    return this.webex.request({
      service: 'hydra',
      resource: 'people/?id=' + ids + '&showAllTypes=' + this.config.showAllTypes
    });
  }
}); /*!
     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
     */

exports.default = PersonUUIDRequestBatcher;
