// @webex/internal-plugin-metrics/dist/call-diagnostic-events-batcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _batcher = require('./batcher');

var _batcher2 = _interopRequireDefault(_batcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CallDiagnosticEventsBatcher = _batcher2.default.extend({
  namespace: 'Metrics',

  prepareItem: function prepareItem(item) {
    // networkType should be a enum value: `wifi`, `ethernet`, `cellular`, or `unknown`.
    // Browsers cannot provide such information right now. However, it is a required field.
    var origin = {
      buildType: process.env.NODE_ENV === 'production' ? 'prod' : 'test',
      networkType: 'unknown'
    };

    item.eventPayload.origin = (0, _assign2.default)(origin, item.eventPayload.origin);

    return _promise2.default.resolve(item);
  },
  prepareRequest: function prepareRequest(queue) {
    // Add sent timestamp
    queue.forEach(function (item) {
      item.eventPayload.originTime.sent = new Date().toISOString();
    });

    return _promise2.default.resolve(queue);
  },
  submitHttpRequest: function submitHttpRequest(payload) {
    return this.webex.request({
      method: 'POST',
      service: 'metrics',
      resource: 'clientmetrics',
      body: {
        metrics: payload
      }
    });
  }
}); /*!
     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
     */

exports.default = CallDiagnosticEventsBatcher;
