// @webex/internal-plugin-metrics/dist/metrics.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _dec, _desc, _value, _obj; /*!
                                * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                                */

var _webexCore = require('@webex/webex-core');

var _common = require('@webex/common');

var _batcher = require('./batcher');

var _batcher2 = _interopRequireDefault(_batcher);

var _clientMetricsBatcher = require('./client-metrics-batcher');

var _clientMetricsBatcher2 = _interopRequireDefault(_clientMetricsBatcher);

var _callDiagnosticEventsBatcher = require('./call-diagnostic-events-batcher');

var _callDiagnosticEventsBatcher2 = _interopRequireDefault(_callDiagnosticEventsBatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Metrics = _webexCore.WebexPlugin.extend((_dec = (0, _common.deprecated)('Metrics#sendUnstructured() is deprecated; please use Metrics#submit()'), (_obj = {
  children: {
    batcher: _batcher2.default,
    clientMetricsBatcher: _clientMetricsBatcher2.default,
    callDiagnosticEventsBatcher: _callDiagnosticEventsBatcher2.default
  },

  namespace: 'Metrics',

  sendUnstructured: function sendUnstructured(key, value) {
    return this.submit(key, value);
  },
  submit: function submit(key, value) {
    return this.batcher.request((0, _assign2.default)({ key: key }, value));
  },


  /**
   * This corresponds to #sendSemiStructured() in the deprecated metrics handler
   * @param {string} eventName
   * @param {Object} props
   * @param {string} preLoginId
   * @returns {Object} HttpResponse object
   */
  submitClientMetrics: function submitClientMetrics(eventName, props, preLoginId) {
    var payload = { metricName: eventName };

    if (props.tags) {
      payload.tags = props.tags;
    }
    if (props.fields) {
      payload.fields = props.fields;
    }
    if (props.type) {
      payload.type = props.type;
    }
    if (props.context) {
      payload.context = props.context;
    }

    if (props.eventPayload) {
      payload.eventPayload = props.eventPayload;
    }

    payload.timestamp = Date.now();

    if (preLoginId) {
      var _payload = {
        metrics: [payload]
      };

      // Do not batch these because pre-login events occur during onboarding, so we will be partially blind
      // to users' progress through the reg flow if we wait to persist pre-login metrics for people who drop off because
      // their metrics will not post from a queue flush in time
      return this.postPreLoginMetric(_payload, preLoginId);
    }

    return this.clientMetricsBatcher.request(payload);
  },


  /**
   * Issue request to alias a user's pre-login ID with their CI UUID
   * @param {string} preLoginId
   * @returns {Object} HttpResponse object
   */
  aliasUser: function aliasUser(preLoginId) {
    return this.request({
      method: 'POST',
      api: 'metrics',
      resource: 'clientmetrics',
      headers: {
        'x-prelogin-userid': preLoginId
      },
      body: {},
      qs: {
        alias: true
      }
    });
  },
  postPreLoginMetric: function postPreLoginMetric(payload, preLoginId) {
    var _this = this;

    return this.webex.credentials.getClientToken().then(function (token) {
      return _this.request({
        method: 'POST',
        api: 'metrics',
        resource: 'clientmetrics-prelogin',
        headers: {
          authorization: token.toString(),
          'x-prelogin-userid': preLoginId
        },
        body: payload
      });
    });
  },
  submitCallDiagnosticEvents: function submitCallDiagnosticEvents(payload) {
    var event = {
      type: 'diagnostic-event',
      eventPayload: payload
    };

    return this.callDiagnosticEventsBatcher.request(event);
  },
  version: '1.72.6'
}, (_applyDecoratedDescriptor(_obj, 'sendUnstructured', [_dec], (0, _getOwnPropertyDescriptor2.default)(_obj, 'sendUnstructured'), _obj)), _obj)));

exports.default = Metrics;
