// @webex/plugin-meetings/dist/peer-connection-manager/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _bowser = require('bowser');

var _bowser2 = _interopRequireDefault(_bowser);

var _sdpTransform = require('sdp-transform');

var _sdpTransform2 = _interopRequireDefault(_sdpTransform);

var _metrics = require('../metrics');

var _metrics2 = _interopRequireDefault(_metrics);

var _loggerProxy = require('../common/logs/logger-proxy');

var _loggerProxy2 = _interopRequireDefault(_loggerProxy);

var _constants = require('../constants');

var _config = require('../metrics/config');

var _media = require('../common/errors/media');

var _media2 = _interopRequireDefault(_media);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @export
 * @public
 */
var pc = {};

/**
 * munges the bandwidth limit into the sdp 
 * @param {String} sdpLines 
 * @param {Number} index
 * @returns {String}
 */
// https://github.com/clux/sdp-transform

/* eslint-disable */
// We need to figure out how to pass a webex logger instance to these util files

/* globals window, RTCSessionDescription */

var insertBandwidthLimit = function insertBandwidthLimit(sdpLines, index) {
  // eslint-disable-next-line no-warning-comments
  // TODO convert to sdp parser
  var limit = void 0;
  if (sdpLines[index].search(_constants.AUDIO) !== -1) {
    limit = _constants.AUDIO_BANDWIDTH_LIMIT;
  } else {
    limit = _constants.VIDEO_BANDWIDTH_LIMIT;
  }
  sdpLines.splice(index + 1, 0, _constants.SDP.B_LINE + ':' + limit);
  return sdpLines;
};

/**
 * needed for calliope max-fs
 * @param {String} sdp 
 * @returns {String}
 */
var setMaxFs = function setMaxFs(sdp) {
  // eslint-disable-next-line no-warning-comments
  // TODO convert with sdp parser, no munging
  var replaceSdp = sdp;
  var payloadTypes = sdp.match(/\na=fmtp:(\d+).*level-asymmetry-allowed=1.*/g);
  if (payloadTypes) {
    payloadTypes.forEach(function (payload) {
      replaceSdp = replaceSdp.replace(payload, payload + ';' + _constants.SDP.MAX_FS_8160);
    });
  }
  return replaceSdp;
};

/**
 * checks that sdp has h264 codec in it
 * @param {String} sdp 
 * @returns {boolean}
 */
var checkH264Support = function checkH264Support(sdp) {
  // eslint-disable-next-line no-warning-comments
  // TODO convert to sdp parser to read rtp.codec
  var videoPresent = sdp.match(/\nm=video.*/g);
  var h264Present = sdp.match(/\na=rtpmap:\d+\sH264.*/g);
  if (videoPresent) {
    return !!h264Present;
  }
  return true;
};

/**
 * validates the sdp, checks port, candidates, and ice info
 * @param {String} sdp
 * @returns {Boolean}
 */
var validateSdp = function validateSdp(sdp) {
  var parsedSdp = _sdpTransform2.default.parse(sdp);
  parsedSdp.media.forEach(function (mediaLine) {
    if (mediaLine.candidates && mediaLine.candidates.length === 0) {
      _loggerProxy2.default.logger.error('PeerConnectionManager->validateSdp#iceCandidate: Ice candadate never completed');
      return false;
    }
    if (_constants.SDP.BAD_MEDIA_PORTS.includes(mediaLine.port)) {
      _loggerProxy2.default.logger.error('PeerConnectionManager->validateSdp#iceCandidate: Found invalid port number for the ice candidate');
      return false;
    }
    if (!mediaLine.icePwd || !mediaLine.iceUfrag) {
      _loggerProxy2.default.logger.error('PeerConnectionManager->validateSdp#iceCandidate: ice ufrag and password not found');
      return false;
    }
    return true;
  });
};

/**
 * munges the bandwidth into the sdp
 * @param {String} sdp 
 * @returns {String}
 */
var limitBandwidth = function limitBandwidth(sdp) {
  // TODO convert to sdp parser
  var offerSdp = sdp;
  var sdpLines = offerSdp.split(_constants.SDP.CARRIAGE_RETURN);

  for (var i = 0; i < sdpLines.length; i += 1) {
    if (sdpLines[i].search(_constants.SDP.M_LINE) !== -1) {
      sdpLines = insertBandwidthLimit(sdpLines, i);
    }
  }
  offerSdp = sdpLines.join(_constants.SDP.CARRIAGE_RETURN);
  return offerSdp;
};

/**
 * sets up a listener for ice fails mid meeting
 * @param {RTCPeerConnection} peerConnection
 * @param {Function} fn
 * @param {Function} name - the name for logging, on the transceiver
 * @returns {undefined}
 */
pc.iceFailListener = function (peerConnection, fn, name) {
  if (!peerConnection) {
    return;
  }
  peerConnection.oniceconnectionstatechange = function () {
    switch (peerConnection.iceConnectionState) {
      case _constants.ICE_STATE.FAILED:
      case _constants.ICE_STATE.DISCONNECTED:
        _loggerProxy2.default.logger.warn('PeerConnectionManager->oniceconnectionstatechange#' + name + ' Interactive Connectivity Establishment(ICE)' + peerConnection.iceConnectionState + ' at ' + new Date());
        setTimeout(function () {
          switch (peerConnection.iceConnectionState) {
            case _constants.ICE_STATE.FAILED:
            case _constants.ICE_STATE.DISCONNECTED:
              // if ice is still failed after some time, do the reconnect
              // else the browser resolves the ice failure on its own
              fn();
              break;
            default:
              break;
          }
        }, _constants.ICE_FAIL_TIMEOUT);
        break;
      default:
        break;
    }
  };
};

/**
 * waits for failures and calls the reconnect function on failures
 * @param {RTCPeerConnection} peerConnection
 * @param {Function} [reconnectFn] - if not included, just prints a log
 * @returns {undefined}
 */
pc.detectFailures = function (peerConnection) {
  var reconnectFn = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : function () {
    _loggerProxy2.default.logger.error('PeerConnectionManager->reconnectFn#Ice failed, no function to reconnect with.');
  };

  peerConnection.iceFailListener(peerConnection, reconnectFn, _constants.MEDIA_PEER_CONNECTION_NAME);
};

/**
 * makes sure the screen pc sdp has content:slides for server
 * @param {RTCPeerConnection} screenPC
 * @returns {RTCPeerConnection}
 */
pc.setContentSlides = function (screenPc) {
  if (screenPc && screenPc.sdp) {
    screenPc.sdp += '' + _constants.SDP.A_CONTENT_SLIDES + _constants.SDP.CARRIAGE_RETURN;
  }
  return screenPc;
};

/**
 * handles ice trickling and establishes ICE connection onto peer connection object
 * @param {Object} peerConnection
 * @returns {RTCPeerConnection}
 */
pc.iceCandidate = function (peerConnection) {
  return new _promise2.default(function (resolve) {
    // TODO: we dont need timeout as we can check the api state and validate.
    var timeout = setTimeout(function () {
      peerConnection.sdp = limitBandwidth(peerConnection.localDescription.sdp);
      peerConnection.sdp = setMaxFs(peerConnection.sdp);
      peerConnection.sdp = peerConnection.sdp.replace(/\na=extmap.*/g, '');

      if (!validateSdp(peerConnection.sdp)) {
        setTimeout(function () {
          resolve(peerConnection);
        }, _constants.RETRY_TIMEOUT);
      } else {
        resolve(peerConnection);
      }
    }, _constants.ICE_TIMEOUT);

    peerConnection.onicecandidate = function (evt) {
      if (!evt.candidate && !peerConnection.sdp) {
        peerConnection.sdp = limitBandwidth(peerConnection.localDescription.sdp);
        peerConnection.sdp = setMaxFs(peerConnection.sdp);
        peerConnection.sdp = peerConnection.sdp.replace(/\na=extmap.*/g, '');

        if (evt.candidate === null) {
          validateSdp(peerConnection.sdp);
          clearTimeout(timeout);
          resolve(peerConnection);
        }
      }
    };
  });
};

/**
 * swapping tracks
 * @param {Object} peerConnection
 * @param {Object} track
 * @returns {undefined}
 */
pc.replaceTrack = function (peerConnection, track) {
  try {
    var senders = peerConnection.getSenders();
    if (senders.length > 0) {
      senders.forEach(function (sender) {
        if (sender.track && sender.track.kind === track.kind) {
          sender.replaceTrack(track);
        }
      });
    }
  } catch (err) {
    _loggerProxy2.default.logger.error('PeerConnectionManager->replaceTrack#Error replacing track, ' + err);
  }
};

/**
 * adding streams to peerConnection
 * @param {Object} peerConnection
 * @param {Object} stream
 * @returns {undefined}
 */
pc.addStream = function (peerConnection, stream) {
  try {
    if (stream && !_bowser2.default.msedge) {
      var tracksPresent = peerConnection.getSenders && peerConnection.getSenders().find(function (sender) {
        return sender.track != null;
      });
      if (tracksPresent) {
        stream.getTracks().forEach(function (track) {
          pc.replaceTrack(peerConnection, track);
        });
        return;
      }
      stream.getTracks().forEach(function (track) {
        peerConnection.addTrack(track, stream);
      });
      // // TODO : may come back disable addTracks for chrome they are moving back to addStream
      // // https://bugs.chromium.org/p/chromium/issues/detail?id=764414
      // // https://bugs.chromium.org/p/chromium/issues/detail?id=738918#c7
      //   peerConnection.addStream(stream);
    } else if (_bowser2.default.msedge) {
      peerConnection.addStream(stream);
    } else {}
  } catch (err) {
    _loggerProxy2.default.logger.error('peer-connection-manager->addStream#Error adding stream, error: ' + _config.error);
  }
};

/**
 * setting the remote description
 * @param {Object} peerConnection
 * @param {String} typeStr
 * @param {String} remoteSdp
 * @param {String} meetingId
 * @returns {undefined}
 */
pc.setRemoteSessionDetails = function (peerConnection, typeStr, remoteSdp, meetingId) {
  _loggerProxy2.default.logger.log('PeerConnectionManager->setRemoteSessionDetails#Setting the remote description type: ' + typeStr + 'State: ' + peerConnection.signalingState);
  var sdp = remoteSdp;

  sdp = sdp.replace(/\na=extmap.*/g, '');

  // making sure that the remoteDescription is only set when there is a answer for offer
  // or there is a offer from the server

  if (!sdp) {
    _metrics2.default.postEvent({
      event: _config.eventType.REMOTE_SDP_RECEIVED,
      meetingId: meetingId, data: {
        canProceed: false,
        errors: [_metrics2.default.generateErrorPayload(2001, true, _config.error.name.MEDIA_ENGINE, 'missing remoteSdp')]
      } });
  }
  if (peerConnection.signalingState === _constants.SDP.HAVE_LOCAL_OFFER || peerConnection.signalingState === _constants.SDP.STABLE && typeStr === _constants.SDP.OFFER) {
    return peerConnection.setRemoteDescription(new window.RTCSessionDescription({
      type: typeStr,
      sdp: sdp
    })).then(function () {
      return _metrics2.default.postEvent({
        event: _config.eventType.REMOTE_SDP_RECEIVED,
        meetingId: meetingId });
    }).catch(function (error) {
      return _metrics2.default.postEvent({
        event: _config.eventType.REMOTE_SDP_RECEIVED,
        meetingId: meetingId,
        data: {
          canProceed: false,
          errors: [_metrics2.default.generateErrorPayload(2001, true, error.name.MEDIA_ENGINE, 'missing remoteSdp')]
        } });
    });
  }

  return _promise2.default.reject(new _media2.default('PeerConnection in wrong state'));
};

/**
 * create offer with a valid paramater
 * @param {Object} params
 * @returns {RTCPeerConnection}
 */
pc.createOffer = function (peerConnection, meetingId) {
  _loggerProxy2.default.logger.log('PeerConnectionManager->createOffer#createOffer: creating a new offer');

  try {
    return peerConnection.createOffer().then(function (description) {
      return peerConnection.setLocalDescription(description);
    }).then(function () {
      return pc.iceCandidate(peerConnection);
    }).then(function () {
      peerConnection.sdp = limitBandwidth(peerConnection.localDescription.sdp);
      peerConnection.sdp = setMaxFs(peerConnection.sdp);
      if (!checkH264Support(peerConnection.sdp)) {
        throw new _media2.default('openH264 is downloading please Wait. Upload logs if not working on second try');
      }

      peerConnection.sdp = peerConnection.sdp.replace(/\na=extmap.*/g, '');
      pc.setContentSlides(peerConnection);

      _metrics2.default.postEvent({
        event: _config.eventType.LOCAL_SDP_GENERATED,
        meetingId: meetingId });
      return peerConnection;
    }).catch(function (err) {
      _metrics2.default.postEvent({
        event: _config.eventType.LOCAL_SDP_GENERATED,
        meetingId: meetingId,
        data: {
          canProceed: false,
          errors: [_metrics2.default.generateErrorPayload(2001, true, _config.error.name.MEDIA_ENGINE)]
        } });
      pc.close(peerConnection);
      throw err;
    });
  } catch (err) {
    throw err;
  }
};

/**
 * rollBack local description in peerconnection
 * @param {Object} peerConnection
 * @returns {RTCPeerConnection}
 */
pc.rollBackLocalDescription = function (peerConnection) {
  return peerConnection.setLocalDescription(new RTCSessionDescription({ type: _constants.SDP.ROLLBACK })).then(function () {
    return peerConnection;
  }).catch(function (err) {
    return _promise2.default.error(err);
  });
};

/**
 * @param {Object} params {
 *  offerToReceiveAudio: {Boolean} sdp constraints
 *  offerToReceiveVideo: {Boolean} sdp constraints
 *  offerSdp: {sdp}
 *  stream: {pcStream}
 * }
 * @returns {Array} [MediaSDP, ScreenSDP]
 */
pc.updatePeerConnection = function (params) {
  _loggerProxy2.default.logger.log('PeerConnectionManager->updatePeerConnection#updating the peerConnection with params: ' + params);

  var _params$meeting$media = params.meeting.mediaProperties,
      peerConnection = _params$meeting$media.peerConnection,
      mediaSettings = _params$meeting$media.mediaSettings;


  return pc.createAnswer({
    peerConnection: peerConnection,
    offerSdp: params.offerSdp[0]
  }).then(function (peerconnection) {
    if (mediaSettings.receiveShare || localShare) {
      pc.setContentSlides(peerconnection);
    }
    return [peerconnection.sdp];
  });
};

/**
 * @param {Object} params {
 *  peerConnection: {Object} peerConnection
 *  sdpConstraints: {Object} sdp constraints
 * }
 * @returns {RTCPeerConnection} peerConnection
 */
pc.createAnswer = function (params) {
  var peerConnection = params.peerConnection;

  // TODO: Some times to many mercury event comes at the same time
  // Need to maintain state of peerconnection

  if (peerConnection.signalingState === 'have-remote-offer') {
    return _promise2.default.resolve(peerConnection);
  }

  return pc.setRemoteSessionDetails(peerConnection, 'offer', params.offerSdp).then(function () {
    return peerConnection.createAnswer(params.sdpConstraints);
  }).then(function (answer) {
    return (
      //   if (edonusHack()) {
      //     answer.sdp = answer.sdp.replace(/recvonly.*/g, 'sendrecv');
      //   }
      peerConnection.setLocalDescription(answer)
    );
  }).then(function () {
    return pc.iceCandidate(peerConnection);
  }).then(function () {
    peerConnection.sdp = limitBandwidth(peerConnection.localDescription.sdp);
    peerConnection.sdp = setMaxFs(peerConnection.sdp);
    if (!checkH264Support(peerConnection.sdp)) {
      throw new _media2.default('openH264 is downloading please Wait. Upload logs if not working on second try');
    }

    peerConnection.sdp = peerConnection.sdp.replace(/\na=extmap.*/g, '');

    return peerConnection;
  }).catch(function (err) {
    _loggerProxy2.default.logger.error('peer-connection-manager->setRemoteSessionDetails#Error creating remote session, error: ' + _config.error);
  });
};

/**
 * shut down the peer connection
 * @param {Object} peerConnection
 * @returns {undefined}
 */
pc.close = function (peerConnection) {
  // peerConnection.close() fails on firefox on network changes and gives a Dom exception
  // To avoid this we have added a try catch block.
  // Please refer to https://bugzilla.mozilla.org/show_bug.cgi?id=1274407 for more information

  if (peerConnection && peerConnection.connectionState === _constants.PEER_CONNECTION_STATE.CLOSED) {
    return _promise2.default.resolve();
  }
  _loggerProxy2.default.logger.log('PeerConnectionManager->close#pc: close() -> closing the mediaPeerConnection');
  return _promise2.default.resolve().then(function () {
    if (peerConnection && peerConnection.close) {
      peerConnection.close();
    }
  });
};

exports.default = pc;
