// @webex/webex-core/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.config = exports.Page = exports.Batcher = exports.RateLimitInterceptor = exports.WebexUserAgentInterceptor = exports.WebexTrackingIdInterceptor = exports.RequestTimingInterceptor = exports.RequestLoggerInterceptor = exports.RequestEventInterceptor = exports.ResponseLoggerInterceptor = exports.RedirectInterceptor = exports.PayloadTransformerInterceptor = exports.NetworkTimingInterceptor = exports.AuthInterceptor = exports.WebexPlugin = exports.StatelessWebexPlugin = exports.WebexHttpError = exports.registerInternalPlugin = exports.registerPlugin = exports.default = exports.children = exports.waitForValue = exports.StorageError = exports.persist = exports.NotFoundError = exports.MemoryStoreAdapter = exports.makeWebexPluginStore = exports.makeWebexStore = exports.Token = exports.sortScope = exports.grantErrors = exports.filterScope = exports.Credentials = undefined;

var _credentials = require('./lib/credentials');

Object.defineProperty(exports, 'Credentials', {
  enumerable: true,
  get: function get() {
    return _credentials.Credentials;
  }
});
Object.defineProperty(exports, 'filterScope', {
  enumerable: true,
  get: function get() {
    return _credentials.filterScope;
  }
});
Object.defineProperty(exports, 'grantErrors', {
  enumerable: true,
  get: function get() {
    return _credentials.grantErrors;
  }
});
Object.defineProperty(exports, 'sortScope', {
  enumerable: true,
  get: function get() {
    return _credentials.sortScope;
  }
});
Object.defineProperty(exports, 'Token', {
  enumerable: true,
  get: function get() {
    return _credentials.Token;
  }
});

var _storage = require('./lib/storage');

Object.defineProperty(exports, 'makeWebexStore', {
  enumerable: true,
  get: function get() {
    return _storage.makeWebexStore;
  }
});
Object.defineProperty(exports, 'makeWebexPluginStore', {
  enumerable: true,
  get: function get() {
    return _storage.makeWebexPluginStore;
  }
});
Object.defineProperty(exports, 'MemoryStoreAdapter', {
  enumerable: true,
  get: function get() {
    return _storage.MemoryStoreAdapter;
  }
});
Object.defineProperty(exports, 'NotFoundError', {
  enumerable: true,
  get: function get() {
    return _storage.NotFoundError;
  }
});
Object.defineProperty(exports, 'persist', {
  enumerable: true,
  get: function get() {
    return _storage.persist;
  }
});
Object.defineProperty(exports, 'StorageError', {
  enumerable: true,
  get: function get() {
    return _storage.StorageError;
  }
});
Object.defineProperty(exports, 'waitForValue', {
  enumerable: true,
  get: function get() {
    return _storage.waitForValue;
  }
});

var _webexCore = require('./webex-core');

Object.defineProperty(exports, 'children', {
  enumerable: true,
  get: function get() {
    return _webexCore.children;
  }
});
Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webexCore).default;
  }
});
Object.defineProperty(exports, 'registerPlugin', {
  enumerable: true,
  get: function get() {
    return _webexCore.registerPlugin;
  }
});
Object.defineProperty(exports, 'registerInternalPlugin', {
  enumerable: true,
  get: function get() {
    return _webexCore.registerInternalPlugin;
  }
});

var _webexHttpError = require('./lib/webex-http-error');

Object.defineProperty(exports, 'WebexHttpError', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webexHttpError).default;
  }
});

var _statelessWebexPlugin = require('./lib/stateless-webex-plugin');

Object.defineProperty(exports, 'StatelessWebexPlugin', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_statelessWebexPlugin).default;
  }
});

var _webexPlugin = require('./lib/webex-plugin');

Object.defineProperty(exports, 'WebexPlugin', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webexPlugin).default;
  }
});

var _auth = require('./interceptors/auth');

Object.defineProperty(exports, 'AuthInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_auth).default;
  }
});

var _networkTiming = require('./interceptors/network-timing');

Object.defineProperty(exports, 'NetworkTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_networkTiming).default;
  }
});

var _payloadTransformer = require('./interceptors/payload-transformer');

Object.defineProperty(exports, 'PayloadTransformerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_payloadTransformer).default;
  }
});

var _redirect = require('./interceptors/redirect');

Object.defineProperty(exports, 'RedirectInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_redirect).default;
  }
});

var _responseLogger = require('./interceptors/response-logger');

Object.defineProperty(exports, 'ResponseLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_responseLogger).default;
  }
});

var _requestEvent = require('./interceptors/request-event');

Object.defineProperty(exports, 'RequestEventInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_requestEvent).default;
  }
});

var _requestLogger = require('./interceptors/request-logger');

Object.defineProperty(exports, 'RequestLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_requestLogger).default;
  }
});

var _requestTiming = require('./interceptors/request-timing');

Object.defineProperty(exports, 'RequestTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_requestTiming).default;
  }
});

var _webexTrackingId = require('./interceptors/webex-tracking-id');

Object.defineProperty(exports, 'WebexTrackingIdInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webexTrackingId).default;
  }
});

var _webexUserAgent = require('./interceptors/webex-user-agent');

Object.defineProperty(exports, 'WebexUserAgentInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webexUserAgent).default;
  }
});

var _rateLimit = require('./interceptors/rate-limit');

Object.defineProperty(exports, 'RateLimitInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rateLimit).default;
  }
});

var _batcher = require('./lib/batcher');

Object.defineProperty(exports, 'Batcher', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_batcher).default;
  }
});

var _page = require('./lib/page');

Object.defineProperty(exports, 'Page', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_page).default;
  }
});

var _config = require('./config');

Object.defineProperty(exports, 'config', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_config).default;
  }
});

require('./plugins/logger');

require('./lib/credentials');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
