// @webex/plugin-meetings/dist/common/logs/logger-proxy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _loggerConfig = require('./logger-config');

var _loggerConfig2 = _interopRequireDefault(_loggerConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LoggerProxy = {
  logger: {
    log: function log() {
      console.error('LoggerProxy->log#NO LOGGER DEFINED');
    },
    error: function error() {
      console.error('LoggerProxy->error#NO LOGGER DEFINED');
    },
    warn: function warn() {
      console.error('LoggerProxy->warn#NO LOGGER DEFINED');
    },
    trace: function trace() {
      console.error('LoggerProxy->trace#NO LOGGER DEFINED');
    },
    debug: function debug() {
      console.error('LoggerProxy->debug#NO LOGGER DEFINED');
    }
  }
};

LoggerProxy.set = function (logger) {
  if (!_loggerConfig2.default.enable) {
    LoggerProxy.logger = {
      log: function log() {},
      error: function error() {},
      warn: function warn() {},
      trace: function trace() {},
      debug: function debug() {}
    };
  }
  LoggerProxy.logger = logger;
};

LoggerProxy.get = function () {
  return LoggerProxy.logger;
};

exports.default = LoggerProxy;
