// @webex/internal-plugin-user/dist/user-uuid-batcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _webexCore = require('@webex/webex-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
var AbstractUserUUIDRequestBatcher = _webexCore.Batcher.extend({
  namespace: 'User',

  /**
   * @param {string} item
   * @returns {Promise<Object>}
   */
  prepareItem: function prepareItem(item) {
    return _promise2.default.resolve({ email: item });
  },


  /**
   * @param {HttpResponseObject} res
   * @returns {Promise}
   */
  handleHttpSuccess: function handleHttpSuccess(res) {
    var _this = this;

    return _promise2.default.all((0, _keys2.default)(res.body).map(function (email) {
      if (res.body[email] && res.body[email].id) {
        return _this.handleItemSuccess(email, res.body[email]);
      }

      return _this.handleItemFailure(email, res.body[email]);
    }));
  },
  handleItemFailure: function handleItemFailure(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.reject(response);
    });
  },


  /**
   * @param {string} email
   * @param {Object} response
   * @returns {Promise}
   */
  handleItemSuccess: function handleItemSuccess(email, response) {
    return this.getDeferredForResponse(email).then(function (defer) {
      defer.resolve(response);
    });
  },


  /**
   * @param {string} email
   * @returns {Promise<string>}
   */
  fingerprintRequest: function fingerprintRequest(email) {
    return _promise2.default.resolve(email.email || email);
  },


  /**
   * @param {string} email
   * @returns {Promise<string>}
   */
  fingerprintResponse: function fingerprintResponse(email) {
    return _promise2.default.resolve(email);
  }
});

/**
 * @class
 */
/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

var FakeUserUUIDRequestBatcher = AbstractUserUUIDRequestBatcher.extend({
  /**
   * @param {Object} payload
   * @returns {Promise<HttpResponseObject>}
   */
  submitHttpRequest: function submitHttpRequest(payload) {
    return this.webex.request({
      method: 'POST',
      service: 'conversation',
      resource: '/users',
      body: payload
    });
  }
});

/**
 * @class
 */
var RealUserUUIDRequestBatcher = AbstractUserUUIDRequestBatcher.extend({
  /**
   * @param {Object} payload
   * @returns {Promise<HttpResponseObject>}
   */
  submitHttpRequest: function submitHttpRequest(payload) {
    return this.webex.request({
      method: 'POST',
      service: 'conversation',
      resource: '/users',
      body: payload,
      qs: {
        shouldCreateUsers: true
      }
    });
  }
});

/**
 * @class
 */
var UserUUIDBatcher = _webexCore.WebexPlugin.extend({
  children: {
    faker: FakeUserUUIDRequestBatcher,
    creator: RealUserUUIDRequestBatcher
  },

  /**
   * @param {Object} payload
   * @returns {Promise<Object>}
   */
  request: function request(payload) {
    return payload.create ? this.creator.request(payload.email) : this.faker.request(payload.email);
  },
  version: '1.72.6'
});

exports.default = UserUUIDBatcher;
