// @webex/plugin-meetings/dist/personal-meeting-room/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _constants = require('../constants');

var PersonalMeetingRoomUtil = {};

PersonalMeetingRoomUtil.getClaimPmrLink = function (pmrLink) {
  var validator = _constants.VALID_PMR_ADDRESS.test(pmrLink);

  if (validator) {
    return pmrLink;
  }
  validator = pmrLink.match(_constants.VALID_PMR_LINK);
  if (validator) {
    // userId + '@' + site + '.webex.com'
    return validator[4] + '@' + validator[2] + '.' + _constants.WEBEX_DOT_COM;
  }

  return null;
};

PersonalMeetingRoomUtil.getClaimPmrPin = function (hostPin) {
  if (_constants.VALID_PIN.test(hostPin)) {
    return hostPin;
  }

  return null;
};

PersonalMeetingRoomUtil.getClaimedRequestParams = function (link, pin, options) {
  return {
    method: _constants.HTTP_VERBS.POST,
    api: _constants.HECATE,
    resource: '/' + _constants.CMR_MEETINGS + '/' + _constants.CLAIM,
    body: {
      userId: options.userId,
      passcode: pin,
      preferred: options.preferred,
      meetingAddress: link
    }
  };
};

exports.default = PersonalMeetingRoomUtil;
