// @webex/helper-html/dist/html.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filterEscapeSync = exports.filterEscape = exports.filterSync = exports.filter = exports.escapeSync = exports.escape = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _curry2 = require('lodash/curry');

var _curry3 = _interopRequireDefault(_curry2);

var _htmlBase = require('./html-base');

Object.defineProperty(exports, 'escape', {
  enumerable: true,
  get: function get() {
    return _htmlBase.escape;
  }
});
Object.defineProperty(exports, 'escapeSync', {
  enumerable: true,
  get: function get() {
    return _htmlBase.escapeSync;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @param {Object} allowedTags
 * @param {Array<string>} allowedStyles
 * @param {string} html
 * @private
 * @returns {string}
 */
function noop() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return new _promise2.default(function (resolve) {
    resolve(noopSync.apply(undefined, args));
  });
}

/**
 * @param {function} processCallback callback function to do additional
 * processing on node. of the form process(node)
 * @param {Object} allowedTags
 * @param {Array<string>} allowedStyles
 * @param {string} html
 * @private
 * @returns {string}
 */
function noopSync(processCallback, allowedTags, allowedStyles, html) {
  return html;
}

var filter = exports.filter = (0, _curry3.default)(noop, 4);
var filterSync = exports.filterSync = (0, _curry3.default)(noopSync, 4);
var filterEscape = exports.filterEscape = (0, _curry3.default)(noop, 4);
var filterEscapeSync = exports.filterEscapeSync = (0, _curry3.default)(noopSync, 4);
