// @webex/common/dist/uuid-utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.constructHydraId = constructHydraId;
exports.deconstructHydraId = deconstructHydraId;
exports.buildHydraMessageId = buildHydraMessageId;
exports.buildHydraPersonId = buildHydraPersonId;
exports.buildHydraRoomId = buildHydraRoomId;
exports.buildHydraOrgId = buildHydraOrgId;
exports.buildHydraMembershipId = buildHydraMembershipId;
exports.getHydraRoomType = getHydraRoomType;
exports.getHydraFiles = getHydraFiles;

var _base = require('./base64');

var _constants = require('./constants');

var hydraBaseResource = 'ciscospark://us';
var hydraBaseUrl = 'https://api.ciscospark.com/v1';

/**
 * Constructs a Hydra ID for a given UUID and type.
 *
 * @export
 * @param {string} type one of PEOPLE, TEAM, ROOM
 * @param {any} id
 * @returns {string}
 */
function constructHydraId(type, id) {
  return (0, _base.encode)(hydraBaseResource + '/' + type.toUpperCase() + '/' + id);
}

/**
 * Deconstructs a Hydra ID and provides the UUID.
 *
 * @export
 * @param {String} id
 * @returns {String}
 */
function deconstructHydraId(id) {
  var payload = (0, _base.decode)(id).split('/');

  return {
    id: payload.pop(),
    type: payload.pop()
  };
}

/**
 * Constructs a Hydra ID for a message based on internal UUID
 *
 * @export
 * @param {any} uuid
 * @returns {string}
 */
function buildHydraMessageId(uuid) {
  return constructHydraId(_constants.hydraTypes.MESSAGE, uuid);
}

/**
 * Constructs a Hydra ID for a person based on internal UUID
 *
 * @export
 * @param {any} uuid
 * @returns {string}
 */
function buildHydraPersonId(uuid) {
  return constructHydraId(_constants.hydraTypes.PEOPLE, uuid);
}

/**
 * Constructs a Hydra ID for a room based on internal UUID
 *
 * @export
 * @param {any} uuid
 * @returns {string}
 */
function buildHydraRoomId(uuid) {
  return constructHydraId(_constants.hydraTypes.ROOM, uuid);
}

/**
 * Constructs a Hydra ID for an organization based on internal UUID
 *
 * @export
 * @param {any} uuid
 * @returns {string}
 */
function buildHydraOrgId(uuid) {
  return constructHydraId(_constants.hydraTypes.ORGANIZATION, uuid);
}

/**
 * Constructs a Hydra ID for an membership based on an
 * internal UUID for the person, and the space
 *
 * @export
 * @param {any} personUUID
 * @param {any} spaceUUID
 * @returns {string}
 */
function buildHydraMembershipId(personUUID, spaceUUID) {
  return constructHydraId(_constants.hydraTypes.MEMBERSHIP, personUUID + ':' + spaceUUID);
}

/**
 * Returns a Hydra roomType based on conversation tags
 *
 * @export
 * @param {arra} tags
 * @param {any} spaceUUID
 * @returns {string}
 */
function getHydraRoomType(tags) {
  if (tags.includes(_constants.SDK_EVENT.INTERNAL.ACTIVITY_TAG.ONE_ON_ONE)) {
    return _constants.SDK_EVENT.EXTERNAL.SPACE_TYPE.DIRECT;
  }

  return _constants.SDK_EVENT.EXTERNAL.SPACE_TYPE.GROUP;
}

/**
 * Returns file URLs for the activity, adhering to Hydra details,
 * e.g., https://api.ciscospark.com/v1/contents/Y2lzY29zcGF...
 * @see https://developer.webex.com/docs/api/v1/messages/get-message-details
 * @param {Object} activity from mercury
 * @returns {Array} file URLs
 */
function getHydraFiles(activity) {
  var hydraFiles = [];
  var files = activity.object.files;


  if (files) {
    var items = files.items;

    // Note: Generated ID is dependent on file order.

    for (var i = 0; i < items.length; i += 1) {
      var contentId = constructHydraId(_constants.hydraTypes.CONTENT, activity.id + '/' + i);

      hydraFiles.push(hydraBaseUrl + '/contents/' + contentId);
    }
  }

  return hydraFiles;
}
